// allure/dist/commands/dashboard.js
import * as console from "node:console";
import { existsSync } from "node:fs";
import { realpath } from "node:fs/promises";
import process, { exit } from "node:process";
import { AllureReport, readConfig } from "@allurereport/core";
import DashboardPlugin from "@allurereport/plugin-dashboard";
import { Command, Option } from "clipanion";
import { red } from "yoctocolors";
export class DashboardCommand extends Command {
    constructor() {
        super(...arguments);
        this.resultsDir = Option.String({ required: true, name: "The directory with Allure results" });
        this.config = Option.String("--config,-c", {
            description: "The path Allure config file",
        });
        this.cwd = Option.String("--cwd", {
            description: "The working directory for the command to run (default: current working directory)",
        });
        this.output = Option.String("--output,-o", {
            description: "The output directory name. Absolute paths are accepted as well",
        });
        this.reportName = Option.String("--report-name,--name", {
            description: "The report name",
        });
        this.singleFile = Option.Boolean("--single-file", {
            description: "Generate single file report",
        });
        this.logo = Option.String("--logo", {
            description: "Path to the report logo which will be displayed in the header",
        });
        this.theme = Option.String("--theme", {
            description: "Default theme of the report (default: OS theme)",
        });
        this.reportLanguage = Option.String("--report-language,--lang", {
            description: "Default language of the report (default: OS language)",
        });
    }
    async execute() {
        if (!existsSync(this.resultsDir)) {
            console.error(red(`The given test results directory doesn't exist: ${this.resultsDir}`));
            exit(1);
            return;
        }
        const cwd = await realpath(this.cwd ?? process.cwd());
        const before = new Date().getTime();
        const defaultDashboardOptions = {
            singleFile: this.singleFile ?? false,
            logo: this.logo,
            theme: this.theme,
            reportLanguage: this.reportLanguage ?? "en",
        };
        const config = await readConfig(cwd, this.config, {
            output: this.output,
            name: this.reportName,
        });
        config.plugins = [
            {
                id: "dashboard",
                enabled: true,
                options: defaultDashboardOptions,
                plugin: new DashboardPlugin(defaultDashboardOptions),
            },
        ];
        const allureReport = new AllureReport(config);
        await allureReport.start();
        await allureReport.readDirectory(this.resultsDir);
        await allureReport.done();
        const after = new Date().getTime();
        console.log(`the report successfully generated (${after - before}ms)`);
    }
}
DashboardCommand.paths = [["dashboard"]];
DashboardCommand.usage = Command.Usage({
    category: "Reports",
    description: "Generates Allure Dashboard report based on provided Allure Results",
    details: "This command generates an Allure Dashboard report from the provided Allure Results directory.",
    examples: [
        ["dashboard ./allure-results", "Generate a report from the ./allure-results directory"],
        [
            "dashboard ./allure-results --output custom-report",
            "Generate a report from the ./allure-results directory to the custom-report directory",
        ],
    ],
});
