// allure/dist/commands/open.js
import { existsSync } from "node:fs";
import { mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { cwd as processCwd } from "node:process";
import { readConfig } from "@allurereport/core";
import { serve } from "@allurereport/static-server";
import { Command, Option } from "clipanion";
import { glob } from "glob";
import { generate } from "./commons/generate.js";
export class OpenCommand extends Command {
    constructor() {
        super(...arguments);
        this.resultsDir = Option.String({
            name: "A report to open or a pattern to match test results directories in the current working directory (default: ./allure-results)",
            required: false,
        });
        this.config = Option.String("--config,-c", {
            description: "The path Allure config file",
        });
        this.port = Option.String("--port", {
            description: "The port to serve the reports on. If not set, the server starts on a random port",
        });
        this.cwd = Option.String("--cwd", {
            description: "The working directory for the command to run (default: current working directory)",
        });
        this.hideLabels = Option.Array("--hide-labels", {
            description: "Hide labels by exact name in generated reports. Repeat the option for multiple labels",
        });
    }
    async execute() {
        const cwd = this.cwd ?? processCwd();
        const hideLabels = this.hideLabels?.length ? this.hideLabels : undefined;
        const targetFullPath = join(cwd, this.resultsDir ?? "allure-report");
        const summaryFiles = existsSync(targetFullPath)
            ? await glob(join(targetFullPath, "**", "summary.json"), {
                mark: true,
                nodir: false,
                absolute: true,
                dot: true,
                windowsPathsNoEscape: true,
                cwd,
            })
            : [];
        if (summaryFiles.length > 0) {
            const config = await readConfig(cwd, this.config, {
                port: this.port,
            });
            await serve({
                port: config.port ? parseInt(config.port, 10) : undefined,
                servePath: targetFullPath,
                open: true,
            });
        }
        else {
            const tmpDir = await mkdtemp(join(tmpdir(), "allure-report-"));
            const config = await readConfig(cwd, this.config, {
                port: this.port,
                output: tmpDir,
                hideLabels,
            });
            await generate({
                resultsDir: targetFullPath,
                cwd,
                config,
            });
            process.on("SIGINT", async () => {
                try {
                    await rm(config.output, { recursive: true });
                }
                catch { }
                process.exit(0);
            });
            await serve({
                port: config.port ? parseInt(config.port, 10) : undefined,
                servePath: config.output,
                open: true,
            });
        }
    }
}
OpenCommand.paths = [["open"], ["serve"]];
OpenCommand.usage = Command.Usage({
    description: "Serves specified directory",
    details: "This command generates report with the given test results and opens it in the default browser.",
    examples: [
        ["open ./allure-results", "Generate and serve the report based on given test results directory"],
        ["open --port 8080 ./allure-report", "Serve the report on port 8080"],
    ],
});
