// allure/dist/commands/whoami.js
import * as console from "node:console";
import { exit } from "node:process";
import { readConfig } from "@allurereport/core";
import { AllureServiceClient, KnownError } from "@allurereport/service";
import { Command, Option } from "clipanion";
import { green, red } from "yoctocolors";
import { logError } from "../utils/logs.js";
export class WhoamiCommand extends Command {
    constructor() {
        super(...arguments);
        this.config = Option.String("--config,-c", {
            description: "The path Allure config file",
        });
        this.cwd = Option.String("--cwd", {
            description: "The working directory for the command to run (default: current working directory)",
        });
    }
    async execute() {
        const config = await readConfig(this.cwd, this.config);
        if (!config?.allureService?.accessToken) {
            console.error(red("No Allure Service access token is provided. Please provide it in the `allureService.accessToken` field in the `allure.config.js` file"));
            exit(1);
            return;
        }
        const serviceClient = new AllureServiceClient(config.allureService);
        const outputLines = [];
        try {
            const { user, project } = await serviceClient.profile();
            outputLines.push(`You are logged in as "${user.email}"`);
            outputLines.push(`Current project is "${project.name}"`);
        }
        catch (error) {
            if (error instanceof KnownError) {
                console.error(red(error.message));
                exit(1);
                return;
            }
            await logError("Failed to get profile due to unexpected error", error);
            exit(1);
        }
        console.info(green(outputLines.join("\n")));
    }
}
WhoamiCommand.paths = [["whoami"]];
WhoamiCommand.usage = Command.Usage({
    category: "Allure Service",
    description: "Prints information about current project",
    details: "This command prints information about the current project based on the Allure Service provided access token.",
    examples: [["whoami", "Print information about the current project using the default configuration"]],
});
