// @allurereport/web-commons/dist/stores/loadableStore/store.js
import { batch, computed, signal } from "@preact/signals-core";
import { LOADABLE_STORE_BRAND } from "./constants.js";
export const loadableStore = (options) => {
    const loadingSignal = signal(false);
    const errorSignal = signal(undefined);
    const dataSignal = signal(options.initialState);
    return {
        value: {
            data: computed(() => dataSignal.value),
            loading: computed(() => loadingSignal.value),
            error: computed(() => errorSignal.value),
            errorMessage: computed(() => {
                if (errorSignal.value) {
                    return errorSignal.value.message;
                }
                return undefined;
            }),
        },
        onInit: () => {
            batch(() => {
                loadingSignal.value = false;
                errorSignal.value = undefined;
                dataSignal.value = options.initialState;
            });
        },
        onLoad: (silent = false) => {
            batch(() => {
                loadingSignal.value = !silent;
                errorSignal.value = undefined;
            });
        },
        onSuccess: (data) => {
            batch(() => {
                loadingSignal.value = false;
                errorSignal.value = undefined;
                dataSignal.value = data;
            });
        },
        onError: (error = new Error("Unknown error")) => {
            batch(() => {
                loadingSignal.value = false;
                errorSignal.value = error;
            });
        },
        brand: LOADABLE_STORE_BRAND,
    };
};
