// @allurereport/core/dist/config.js
import * as console from "node:console";
import { readFile, stat } from "node:fs/promises";
import { extname, resolve } from "node:path";
import * as process from "node:process";
import { validateEnvironmentName } from "@allurereport/core-api";
import { parse } from "yaml";
import { readKnownIssues } from "./known.js";
import { FileSystemReportFiles } from "./plugin.js";
import { environmentIdentityById, environmentIdentityByName, validateAllowedEnvironmentIds, normalizeEnvironmentDescriptorMap, validateAllowedEnvironmentId, } from "./utils/environment.js";
import { importWrapper } from "./utils/module.js";
import { normalizeImportPath } from "./utils/path.js";
import { assertValidPluginIdForWindows, isWindows } from "./utils/windows.js";
const CONFIG_FILENAMES = [
    "allurerc.js",
    "allurerc.mjs",
    "allurerc.cjs",
    "allurerc.json",
    "allurerc.yaml",
    "allurerc.yml",
];
const DEFAULT_CONFIG = {};
const isAgentDescriptor = (value) => {
    return value === "agent" || value === "@allurereport/plugin-agent";
};
const hasConfiguredAgent = (plugins) => {
    return Object.entries(plugins).some(([key, descriptor]) => isAgentDescriptor(key) || isAgentDescriptor(descriptor.import));
};
export const assertValidPluginId = (id) => {
    if (id.length === 0) {
        throw new Error("Invalid plugin id: must not be empty");
    }
    if (id === "." || id === "..") {
        throw new Error(`Invalid plugin id ${JSON.stringify(id)}: must not be "." or ".."`);
    }
    if (id.includes("..")) {
        throw new Error(`Invalid plugin id ${JSON.stringify(id)}: must not contain ".."`);
    }
    if (/[/\\]/.test(id)) {
        throw new Error(`Invalid plugin id ${JSON.stringify(id)}: must not contain path separators`);
    }
    if (isWindows()) {
        assertValidPluginIdForWindows(id);
    }
};
export const getPluginId = (key) => {
    const trimmed = key.trim();
    if (trimmed.length === 0) {
        throw new Error(`Invalid plugin key ${JSON.stringify(key)}: must not be empty or whitespace-only`);
    }
    const id = trimmed.replace(/^@.*\//, "").replace(/[/\\]/g, "-");
    assertValidPluginId(id);
    return id;
};
export const findConfig = async (cwd, configPath) => {
    if (configPath) {
        const resolved = resolve(cwd, configPath);
        try {
            const stats = await stat(resolved);
            if (stats.isFile()) {
                return resolved;
            }
        }
        catch (e) {
            console.error(e);
        }
        throw new Error(`invalid config path ${resolved}: not a regular file`);
    }
    for (const configFilename of CONFIG_FILENAMES) {
        const resolved = resolve(cwd, configFilename);
        try {
            const stats = await stat(resolved);
            if (stats.isFile()) {
                return resolved;
            }
        }
        catch (ignored) {
        }
    }
};
export const validateConfig = (config) => {
    const supportedFields = [
        "name",
        "output",
        "open",
        "port",
        "hideLabels",
        "historyPath",
        "historyLimit",
        "knownIssuesPath",
        "plugins",
        "defaultLabels",
        "variables",
        "environment",
        "allowedEnvironments",
        "environments",
        "appendHistory",
        "qualityGate",
        "allureService",
        "categories",
        "globalAttachments",
    ];
    const unsupportedFields = Object.keys(config).filter((key) => !supportedFields.includes(key));
    return {
        valid: unsupportedFields.length === 0,
        fields: unsupportedFields,
    };
};
export const loadYamlConfig = async (configPath) => {
    try {
        const rawConfig = await readFile(configPath, "utf-8");
        const parsedConfig = parse(rawConfig);
        return parsedConfig || DEFAULT_CONFIG;
    }
    catch (err) {
        if (err?.code === "ENOENT") {
            return DEFAULT_CONFIG;
        }
        throw err;
    }
};
export const loadJsonConfig = async (configPath) => {
    try {
        const rawConfig = await readFile(configPath, "utf-8");
        const parsedConfig = JSON.parse(rawConfig);
        return parsedConfig || DEFAULT_CONFIG;
    }
    catch (err) {
        if (err?.code === "ENOENT") {
            return DEFAULT_CONFIG;
        }
        throw err;
    }
};
export const loadJsConfig = async (configPath) => {
    return (await import(normalizeImportPath(configPath))).default;
};
const resolveConfigEnvironments = (config) => {
    const errors = [];
    const { ids: allowedEnvironments, idsSet: allowedEnvironmentIds, errors: allowedEnvironmentErrors, } = validateAllowedEnvironmentIds(config.allowedEnvironments, "config.allowedEnvironments");
    const { normalized: environments, errors: environmentErrors } = normalizeEnvironmentDescriptorMap(config.environments ?? {}, "config.environments");
    let environment;
    errors.push(...allowedEnvironmentErrors, ...environmentErrors);
    if (config.environment !== undefined) {
        const environmentResult = validateEnvironmentName(config.environment);
        if (!environmentResult.valid) {
            errors.push(`environment ${environmentResult.reason}`);
        }
        else {
            const normalizedEnvironment = environmentResult.normalized;
            environment =
                environmentIdentityById(environments, normalizedEnvironment)?.id ??
                    environmentIdentityByName(environments, normalizedEnvironment)?.id ??
                    normalizedEnvironment;
            const allowedEnvironmentError = validateAllowedEnvironmentId(environment, allowedEnvironmentIds, "config");
            if (allowedEnvironmentError) {
                throw new Error(`The provided Allure config contains invalid environments: ${allowedEnvironmentError}`);
            }
        }
    }
    for (const environmentId of Object.keys(environments)) {
        const allowedEnvironmentError = validateAllowedEnvironmentId(environmentId, allowedEnvironmentIds, "config.environments");
        if (allowedEnvironmentError) {
            throw new Error(`The provided Allure config contains invalid environments: ${allowedEnvironmentError}`);
        }
    }
    if (errors.length > 0) {
        throw new Error(`The provided Allure config contains invalid environments: ${errors.join("; ")}`);
    }
    return {
        environments,
        environment,
        allowedEnvironments,
    };
};
export const resolveConfig = async (config, override = {}) => {
    const validationResult = validateConfig(config);
    if (!validationResult.valid) {
        throw new Error(`The provided Allure config contains unsupported fields: ${validationResult.fields.join(", ")}`);
    }
    const { environments, environment, allowedEnvironments } = resolveConfigEnvironments(config);
    const name = override.name ?? config.name ?? "Allure Report";
    const open = override.open ?? config.open ?? false;
    const port = override.port ?? config.port ?? undefined;
    const hideLabels = override.hideLabels ?? config.hideLabels;
    const historyPath = override.historyPath ?? config.historyPath;
    const historyLimit = override.historyLimit ?? config.historyLimit;
    const appendHistory = config.appendHistory ?? true;
    const knownIssuesPath = resolve(override.knownIssuesPath ?? config.knownIssuesPath ?? "./allure/known.json");
    const output = resolve(override.output ?? config.output ?? "./allure-report");
    const known = await readKnownIssues(knownIssuesPath);
    const variables = config.variables ?? {};
    const configuredPlugins = override.plugins ?? config.plugins;
    const basePlugins = Object.keys(configuredPlugins ?? {}).length === 0
        ? {
            awesome: {
                options: {},
            },
        }
        : configuredPlugins;
    const plugins = hasConfiguredAgent(basePlugins)
        ? basePlugins
        : {
            ...basePlugins,
            agent: {
                options: {},
            },
        };
    const pluginInstances = await resolvePlugins(plugins);
    return {
        name,
        output,
        open,
        port,
        hideLabels,
        knownIssuesPath,
        known,
        environment,
        allowedEnvironments,
        variables,
        environments,
        appendHistory,
        historyLimit,
        historyPath: historyPath ? resolve(historyPath) : undefined,
        reportFiles: new FileSystemReportFiles(output),
        plugins: pluginInstances,
        defaultLabels: config.defaultLabels ?? {},
        qualityGate: config.qualityGate,
        allureService: config.allureService,
        categories: config.categories,
        globalAttachments: config.globalAttachments,
    };
};
export const readConfig = async (cwd = process.cwd(), configPath, override) => {
    const cfg = (await findConfig(cwd, configPath)) ?? "";
    let config;
    switch (extname(cfg)) {
        case ".json":
            config = await loadJsonConfig(cfg);
            break;
        case ".yaml":
        case ".yml":
            config = await loadYamlConfig(cfg);
            break;
        case ".js":
        case ".cjs":
        case ".mjs":
            config = await loadJsConfig(cfg);
            break;
        default:
            config = DEFAULT_CONFIG;
    }
    const fullConfig = await resolveConfig(config, override);
    return fullConfig;
};
export const getPluginInstance = (config, predicate) => {
    return config?.plugins?.find(predicate);
};
const isModuleNotFoundError = (err) => {
    return err instanceof Error && "code" in err && err.code === "ERR_MODULE_NOT_FOUND";
};
export const resolvePlugin = async (path) => {
    if (!path.startsWith("@allurereport/plugin-")) {
        try {
            const module = await importWrapper(`@allurereport/plugin-${path}`);
            return module.default;
        }
        catch (err) {
            if (!isModuleNotFoundError(err)) {
                throw err;
            }
        }
    }
    try {
        const module = await importWrapper(path);
        return module.default;
    }
    catch {
        throw new Error(`Cannot resolve plugin: ${path}`);
    }
};
const resolvePlugins = async (plugins) => {
    const pluginInstances = [];
    for (const id in plugins) {
        const pluginConfig = plugins[id];
        const pluginId = getPluginId(id);
        const Plugin = await resolvePlugin(pluginConfig.import ?? id);
        pluginInstances.push({
            id: pluginId,
            enabled: pluginConfig.enabled ?? true,
            options: pluginConfig.options ?? {},
            plugin: new Plugin(pluginConfig.options),
        });
    }
    return pluginInstances;
};
