// @allurereport/plugin-api/dist/store.js
export var AllureStoreDumpFiles;
(function (AllureStoreDumpFiles) {
    AllureStoreDumpFiles["TestResults"] = "test-results.json";
    AllureStoreDumpFiles["TestCases"] = "test-cases.json";
    AllureStoreDumpFiles["Fixtures"] = "fixtures.json";
    AllureStoreDumpFiles["GlobalErrors"] = "global-errors.json";
    AllureStoreDumpFiles["GlobalAttachments"] = "global-attachments.json";
    AllureStoreDumpFiles["Attachments"] = "attachments.json";
    AllureStoreDumpFiles["Environments"] = "environments.json";
    AllureStoreDumpFiles["ReportVariables"] = "report-variables.json";
    AllureStoreDumpFiles["IndexAttachmentsByTestResults"] = "index-attachments-by-test-results.json";
    AllureStoreDumpFiles["IndexTestResultsByHistoryId"] = "index-test-results-by-history-id.json";
    AllureStoreDumpFiles["IndexTestResultsByTestCase"] = "index-test-results-by-test-case.json";
    AllureStoreDumpFiles["IndexLatestEnvTestResultsByHistoryId"] = "index-latest-env-test-results-by-history-id.json";
    AllureStoreDumpFiles["IndexAttachmentsByFixture"] = "index-attachments-by-fixture.json";
    AllureStoreDumpFiles["IndexFixturesByTestResult"] = "index-fixtures-by-test-result.json";
    AllureStoreDumpFiles["IndexKnownByHistoryId"] = "index-known-by-history-id.json";
    AllureStoreDumpFiles["QualityGateResults"] = "quality-gate-results.json";
})(AllureStoreDumpFiles || (AllureStoreDumpFiles = {}));
