// @allurereport/service/dist/service.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AllureServiceClient_client, _AllureServiceClient_url;
import { readFile } from "node:fs/promises";
import { join as joinPosix } from "node:path/posix";
import { createServiceHttpClient } from "./utils/http.js";
const ASSET_MAX_FILE_SIZE = 200 * 1024 * 1024;
export class AllureServiceClient {
    constructor(config) {
        this.config = config;
        _AllureServiceClient_client.set(this, void 0);
        _AllureServiceClient_url.set(this, void 0);
        if (!config?.accessToken) {
            throw new Error("Allure service access token is required");
        }
        const { iss, projectId, url: baseUrl } = this.decodeToken(config.accessToken) ?? {};
        if (iss !== "allure-service" || !baseUrl || !projectId) {
            throw new Error("Invalid access token");
        }
        __classPrivateFieldSet(this, _AllureServiceClient_url, baseUrl, "f");
        __classPrivateFieldSet(this, _AllureServiceClient_client, createServiceHttpClient(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), config.accessToken), "f");
    }
    decodeToken(token) {
        try {
            const base64Url = token.split(".")[1];
            const base64 = base64Url.replace(/-/g, "+").replace(/_/g, "/");
            return JSON.parse(atob(base64));
        }
        catch {
            return undefined;
        }
    }
    async profile() {
        const { user, project } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get("/user/profile");
        return {
            user,
            project,
        };
    }
    async generateNewAccessToken(projectUuid) {
        const { accessToken } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/auth/tokens", {
            body: {
                projectId: projectUuid,
            },
        });
        return accessToken;
    }
    async projects() {
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get("/projects");
    }
    async project(uuid) {
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get(`/projects/${uuid}`);
    }
    async downloadHistory(payload) {
        const { branch, limit } = payload ?? {};
        const params = new URLSearchParams();
        if (limit) {
            params.append("limit", encodeURIComponent(limit));
        }
        if (branch) {
            params.append("branch", encodeURIComponent(branch));
        }
        const { history } = await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").get(params.size > 0 ? `/projects/history?${params.toString()}` : "/projects/history");
        return history;
    }
    async createReport(payload) {
        const { reportName, reportUuid, branch } = payload;
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/reports", {
            body: {
                reportName,
                reportUuid,
                branch,
            },
        });
    }
    async completeReport(payload) {
        const { reportUuid, historyPoint } = payload;
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/reports/${reportUuid}/complete`, {
            body: {
                historyPoint,
            },
        });
    }
    async deleteReport(payload) {
        const { reportUuid, pluginId = "" } = payload;
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/reports/${reportUuid}/delete`, {
            body: {
                pluginId,
            },
        });
    }
    async addReportAsset(payload) {
        const { filename, file, filepath } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = await readFile(filepath);
        }
        if (content.length > ASSET_MAX_FILE_SIZE) {
            throw new Error(`Asset size exceeds the maximum allowed size of ${ASSET_MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }
        const form = new FormData();
        form.set("filename", filename);
        form.set("file", content);
        return __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post("/assets/upload", {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });
    }
    async addReportFile(payload) {
        const { reportUuid, filename, file, filepath, pluginId } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = await readFile(filepath);
        }
        if (content.length > ASSET_MAX_FILE_SIZE) {
            throw new Error(`Report file size exceeds the maximum allowed size of ${ASSET_MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }
        const form = new FormData();
        form.set("filename", pluginId ? joinPosix(pluginId, filename) : filename);
        form.set("file", content);
        await __classPrivateFieldGet(this, _AllureServiceClient_client, "f").post(`/reports/${reportUuid}/upload`, {
            body: form,
            headers: {
                "Content-Type": "multipart/form-data",
            },
        });
        return joinPosix(__classPrivateFieldGet(this, _AllureServiceClient_url, "f"), reportUuid, filename);
    }
}
_AllureServiceClient_client = new WeakMap(), _AllureServiceClient_url = new WeakMap();
