// @allurereport/web-commons/dist/charts/accessors/problemsDistributionHeatMap.js
import { filterIncludedInSuccessRate } from "@allurereport/core-api";
const groupTestsByEnvironment = (testResults) => {
    return testResults.reduce((acc, testResult) => {
        const key = testResult.environment;
        if (key) {
            const bucket = acc[key] || (acc[key] = []);
            bucket.push(testResult);
        }
        return acc;
    }, {});
};
const groupByExactLabel = (testResults, labelNames) => {
    return testResults.reduce((acc, testResult) => {
        const labels = testResult.labels;
        if (!labels) {
            return acc;
        }
        for (const label of labels) {
            const key = label.value;
            if (labelNames.includes(label.name) && key) {
                const bucket = acc[key] || (acc[key] = []);
                bucket.push(testResult);
            }
        }
        return acc;
    }, {});
};
const makeHeatMapSerie = (env, testResults) => {
    const testResultsByExactLabel = groupByExactLabel(testResults, ["feature"]);
    const data = [];
    for (const [labelValue, testsByLabelValue] of Object.entries(testResultsByExactLabel)) {
        const testsTotal = testsByLabelValue.length;
        const totalNegative = testsByLabelValue.reduce((acc, test) => acc + (test.status !== "passed" ? 1 : 0), 0);
        data.push({
            x: labelValue,
            y: totalNegative / testsTotal,
        });
    }
    return {
        id: env,
        data: data.sort((a, b) => (a.y || 0) - (b.y || 0)),
    };
};
const makeHeatMapData = (testsByEnvironment) => {
    return Object.entries(testsByEnvironment).map(([env, tests]) => makeHeatMapSerie(env, tests));
};
const filterTestResultsBySignificantStatus = (testResults) => {
    return testResults.filter(filterIncludedInSuccessRate);
};
const filterTestResultsByLabelNames = (testResults, labelNames) => {
    return testResults.filter((test) => test.labels?.some((l) => labelNames.includes(l.name)));
};
export const problemsDistributionHeatMapAccessor = {
    getHeatMap: ({ testResults }) => {
        const filteredTestResults = filterTestResultsBySignificantStatus(filterTestResultsByLabelNames(testResults, ["feature"]));
        const testsResultsByEnvironment = groupTestsByEnvironment(filteredTestResults);
        const data = makeHeatMapData(testsResultsByEnvironment);
        const totals = new Map();
        for (const serie of data) {
            const total = serie.data.reduce((acc, sd) => acc + (sd.y || 0), 0);
            totals.set(serie.id, total);
        }
        return data.sort((a, b) => totals.get(a.id) - totals.get(b.id));
    },
};
