// @allurereport/core/dist/store/store.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _DefaultAllureStore_instances, _DefaultAllureStore_testResults, _DefaultAllureStore_environmentDisplayNames, _DefaultAllureStore_environmentNameToId, _DefaultAllureStore_attachments, _DefaultAllureStore_attachmentContents, _DefaultAllureStore_testCases, _DefaultAllureStore_metadata, _DefaultAllureStore_history, _DefaultAllureStore_known, _DefaultAllureStore_fixtures, _DefaultAllureStore_defaultLabels, _DefaultAllureStore_environment, _DefaultAllureStore_environmentsConfig, _DefaultAllureStore_reportVariables, _DefaultAllureStore_realtimeDispatcher, _DefaultAllureStore_realtimeSubscriber, _DefaultAllureStore_allowedEnvironmentIds, _DefaultAllureStore_globalAttachmentIds, _DefaultAllureStore_globalAttachmentIdsByEnv, _DefaultAllureStore_globalErrors, _DefaultAllureStore_globalErrorsByEnv, _DefaultAllureStore_globalExitCode, _DefaultAllureStore_qualityGateResults, _DefaultAllureStore_historyPoints, _DefaultAllureStore_environments, _DefaultAllureStore_mergeEnvironmentIdentity, _DefaultAllureStore_environmentIdForLookup, _DefaultAllureStore_addEnvironments, _DefaultAllureStore_environmentIdByName, _DefaultAllureStore_setTestResultEnvironmentId, _DefaultAllureStore_assertAllowedEnvironmentId, _DefaultAllureStore_assertAllowedStoredEnvironment, _DefaultAllureStore_environmentIdByTestResult, _DefaultAllureStore_resolveGlobalEnvironmentIdentity, _DefaultAllureStore_globalAttachmentId, _DefaultAllureStore_indexGlobalError, _DefaultAllureStore_addGlobalError, _DefaultAllureStore_indexGlobalAttachment, _DefaultAllureStore_addGlobalAttachment;
import { extname } from "node:path";
import { DEFAULT_ENVIRONMENT, DEFAULT_ENVIRONMENT_IDENTITY, compareBy, createDictionary, getHistoryIdCandidates, getWorstStatus, nullsLast, ordinal, reverse, selectHistoryTestResults, validateEnvironmentId, validateEnvironmentName, } from "@allurereport/core-api";
import { md5, } from "@allurereport/plugin-api";
import { environmentIdentityById, normalizeEnvironmentDescriptorMap, resolveEnvironmentIdentity, resolveStoredEnvironmentIdentity, validateAllowedEnvironmentId, } from "../utils/environment.js";
import { isFlaky } from "../utils/flaky.js";
import { getStatusTransition } from "../utils/new.js";
import { testFixtureResultRawToState, testResultRawToState } from "./convert.js";
const index = (indexMap, key, ...items) => {
    if (key) {
        if (!indexMap.has(key)) {
            indexMap.set(key, []);
        }
        const current = indexMap.get(key);
        current.push(...items);
    }
};
const wasStartedEarlier = (first, second) => first.start === undefined || second.start === undefined || first.start < second.start;
const hidePreviousAttemptByEnvironment = (state, testResult, environmentId) => {
    const { historyId } = testResult;
    if (environmentId) {
        if (!state.has(environmentId)) {
            state.set(environmentId, new Map());
        }
        if (historyId) {
            const historyIdToLastAttemptResult = state.get(environmentId);
            const currentLastAttemptResult = historyIdToLastAttemptResult.get(historyId);
            if (currentLastAttemptResult) {
                if (wasStartedEarlier(currentLastAttemptResult, testResult)) {
                    historyIdToLastAttemptResult.set(historyId, testResult);
                    currentLastAttemptResult.hidden = true;
                }
                else {
                    testResult.hidden = true;
                }
            }
            else {
                historyIdToLastAttemptResult.set(historyId, testResult);
            }
        }
    }
};
export const mapToObject = (map) => {
    const result = {};
    map.forEach((value, key) => {
        result[key] = value;
    });
    return result;
};
export const updateMapWithRecord = (map, record) => {
    Object.entries(record).forEach(([key, value]) => {
        map.set(key, value);
    });
    return map;
};
export class DefaultAllureStore {
    constructor(params) {
        _DefaultAllureStore_instances.add(this);
        _DefaultAllureStore_testResults.set(this, void 0);
        _DefaultAllureStore_environmentDisplayNames.set(this, void 0);
        _DefaultAllureStore_environmentNameToId.set(this, void 0);
        _DefaultAllureStore_attachments.set(this, void 0);
        _DefaultAllureStore_attachmentContents.set(this, void 0);
        _DefaultAllureStore_testCases.set(this, void 0);
        _DefaultAllureStore_metadata.set(this, void 0);
        _DefaultAllureStore_history.set(this, void 0);
        _DefaultAllureStore_known.set(this, void 0);
        _DefaultAllureStore_fixtures.set(this, void 0);
        _DefaultAllureStore_defaultLabels.set(this, {});
        _DefaultAllureStore_environment.set(this, void 0);
        _DefaultAllureStore_environmentsConfig.set(this, {});
        _DefaultAllureStore_reportVariables.set(this, {});
        _DefaultAllureStore_realtimeDispatcher.set(this, void 0);
        _DefaultAllureStore_realtimeSubscriber.set(this, void 0);
        _DefaultAllureStore_allowedEnvironmentIds.set(this, void 0);
        this.indexTestResultByTestCase = new Map();
        this.indexTestResultByEnvironmentId = new Map();
        this.indexLatestEnvTestResultByHistoryId = new Map();
        this.indexTestResultByHistoryId = new Map();
        this.indexAttachmentByTestResult = new Map();
        this.indexAttachmentByFixture = new Map();
        this.indexFixturesByTestResult = new Map();
        this.indexKnownByHistoryId = new Map();
        _DefaultAllureStore_globalAttachmentIds.set(this, []);
        _DefaultAllureStore_globalAttachmentIdsByEnv.set(this, new Map());
        _DefaultAllureStore_globalErrors.set(this, []);
        _DefaultAllureStore_globalErrorsByEnv.set(this, new Map());
        _DefaultAllureStore_globalExitCode.set(this, void 0);
        _DefaultAllureStore_qualityGateResults.set(this, []);
        _DefaultAllureStore_historyPoints.set(this, []);
        _DefaultAllureStore_environments.set(this, []);
        const { history, known = [], realtimeDispatcher, realtimeSubscriber, defaultLabels = {}, environment, allowedEnvironments, environmentsConfig = {}, reportVariables = {}, } = params ?? {};
        const errors = [];
        const { normalized: normalizedEnvironmentsConfig, identities, errors: configErrors, } = normalizeEnvironmentDescriptorMap(environmentsConfig, "environmentsConfig");
        errors.push(...configErrors);
        const forcedEnvironment = resolveEnvironmentIdentity({
            environment,
        }, normalizedEnvironmentsConfig, "store constructor");
        errors.push(...forcedEnvironment.errors);
        const resolvedEnvironment = forcedEnvironment.identity;
        if (errors.length > 0) {
            throw new Error(errors.join("; "));
        }
        const environments = [...identities];
        if (resolvedEnvironment) {
            environments.push(resolvedEnvironment);
        }
        __classPrivateFieldSet(this, _DefaultAllureStore_testResults, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_environmentDisplayNames, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_environmentNameToId, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_attachments, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_attachmentContents, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_testCases, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_metadata, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_fixtures, new Map(), "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_history, history, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_known, [...known], "f");
        __classPrivateFieldGet(this, _DefaultAllureStore_known, "f").forEach((ktf) => index(this.indexKnownByHistoryId, ktf.historyId, ktf));
        __classPrivateFieldSet(this, _DefaultAllureStore_realtimeDispatcher, realtimeDispatcher, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_realtimeSubscriber, realtimeSubscriber, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_defaultLabels, defaultLabels, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_environmentsConfig, normalizedEnvironmentsConfig, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_environment, resolvedEnvironment, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_reportVariables, reportVariables, "f");
        __classPrivateFieldSet(this, _DefaultAllureStore_allowedEnvironmentIds, new Set(allowedEnvironments ?? []), "f");
        __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addEnvironments).call(this, environments);
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeSubscriber, "f")?.onQualityGateResults(async (results) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_qualityGateResults, "f").push(...results);
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addEnvironments).call(this, results
                .map((result) => {
                const aliasedEnvironmentId = typeof result.environment === "string" ? __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, result.environment) : undefined;
                if (aliasedEnvironmentId) {
                    const aliasedEnvironment = __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").find(({ id }) => id === aliasedEnvironmentId)?.name ?? aliasedEnvironmentId;
                    return {
                        id: aliasedEnvironmentId,
                        name: aliasedEnvironment,
                    };
                }
                return resolveStoredEnvironmentIdentity({
                    environmentName: result.environment,
                    labels: [],
                }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
                    forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
                });
            })
                .filter(Boolean));
        });
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeSubscriber, "f")?.onGlobalExitCode(async (exitCode) => {
            __classPrivateFieldSet(this, _DefaultAllureStore_globalExitCode, exitCode, "f");
        });
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeSubscriber, "f")?.onGlobalError(async (error) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addGlobalError).call(this, error);
        });
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeSubscriber, "f")?.onGlobalAttachment(async ({ attachment, fileName, environment }) => {
            const originalFileName = attachment.getOriginalFileName();
            const resolvedEnvironment = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_resolveGlobalEnvironmentIdentity).call(this, environment);
            const attachmentLink = {
                id: __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_globalAttachmentId).call(this, originalFileName, resolvedEnvironment?.id),
                name: fileName || originalFileName,
                missed: false,
                used: true,
                ext: attachment.getExtension(),
                contentType: attachment.getContentType(),
                contentLength: attachment.getContentLength(),
                originalFileName,
                environment: resolvedEnvironment?.name ?? environment,
            };
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addGlobalAttachment).call(this, attachmentLink, attachment);
        });
    }
    async readHistory() {
        if (!__classPrivateFieldGet(this, _DefaultAllureStore_history, "f")) {
            return [];
        }
        __classPrivateFieldSet(this, _DefaultAllureStore_historyPoints, (await __classPrivateFieldGet(this, _DefaultAllureStore_history, "f").readHistory()) ?? [], "f");
        __classPrivateFieldGet(this, _DefaultAllureStore_historyPoints, "f").sort(compareBy("timestamp", reverse(ordinal())));
        return __classPrivateFieldGet(this, _DefaultAllureStore_historyPoints, "f");
    }
    async appendHistory(history) {
        if (!__classPrivateFieldGet(this, _DefaultAllureStore_history, "f")) {
            return;
        }
        __classPrivateFieldGet(this, _DefaultAllureStore_historyPoints, "f").push(history);
        await __classPrivateFieldGet(this, _DefaultAllureStore_history, "f").appendHistory(history);
    }
    async qualityGateResults() {
        return [...__classPrivateFieldGet(this, _DefaultAllureStore_qualityGateResults, "f")];
    }
    async qualityGateResultsByEnv() {
        const resultsById = await this.qualityGateResultsByEnvironmentId();
        const resultsByEnv = createDictionary();
        Object.entries(resultsById).forEach(([environmentId, results]) => {
            const environmentName = __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").find(({ id }) => id === environmentId)?.name ?? environmentId;
            resultsByEnv[environmentName] = results;
        });
        return resultsByEnv;
    }
    async qualityGateResultsByEnvironmentId() {
        const resultsById = createDictionary();
        for (const result of __classPrivateFieldGet(this, _DefaultAllureStore_qualityGateResults, "f")) {
            const aliasedEnvironmentId = typeof result.environment === "string" ? __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, result.environment) : undefined;
            const envIdentity = aliasedEnvironmentId !== undefined
                ? {
                    id: aliasedEnvironmentId,
                    name: __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").find(({ id }) => id === aliasedEnvironmentId)?.name ?? aliasedEnvironmentId,
                }
                : resolveStoredEnvironmentIdentity({
                    environmentName: result.environment,
                    labels: [],
                }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
                    forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
                });
            if (!envIdentity) {
                continue;
            }
            if (!resultsById[envIdentity.id]) {
                resultsById[envIdentity.id] = [];
            }
            resultsById[envIdentity.id].push(result);
        }
        return resultsById;
    }
    async globalExitCode() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_globalExitCode, "f");
    }
    async allGlobalErrors() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_globalErrors, "f");
    }
    async allGlobalErrorsByEnv() {
        return mapToObject(__classPrivateFieldGet(this, _DefaultAllureStore_globalErrorsByEnv, "f"));
    }
    async allGlobalAttachments() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIds, "f").reduce((acc, id) => {
            const attachment = __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(id);
            if (!attachment) {
                return acc;
            }
            acc.push(attachment);
            return acc;
        }, []);
    }
    async allGlobalAttachmentsByEnv() {
        const result = {};
        __classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIdsByEnv, "f").forEach((attachmentIds, environmentId) => {
            result[environmentId] = attachmentIds.reduce((acc, id) => {
                const attachment = __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(id);
                if (!attachment) {
                    return acc;
                }
                acc.push(attachment);
                return acc;
            }, []);
        });
        return result;
    }
    async visitTestResult(raw, context) {
        const attachmentLinks = [];
        const testResult = testResultRawToState({
            testCases: __classPrivateFieldGet(this, _DefaultAllureStore_testCases, "f"),
            attachments: __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f"),
            visitAttachmentLink: (link) => attachmentLinks.push(link),
        }, raw, context);
        const defaultLabelsNames = Object.keys(__classPrivateFieldGet(this, _DefaultAllureStore_defaultLabels, "f"));
        if (defaultLabelsNames.length) {
            defaultLabelsNames.forEach((labelName) => {
                if (!testResult.labels.find((label) => label.name === labelName)) {
                    const defaultLabelValue = __classPrivateFieldGet(this, _DefaultAllureStore_defaultLabels, "f")[labelName];
                    [].concat(defaultLabelValue).forEach((labelValue) => {
                        testResult.labels.push({
                            name: labelName,
                            value: labelValue,
                        });
                    });
                }
            });
        }
        const environmentIdentity = __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f") ??
            (() => {
                const match = Object.entries(__classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f")).find(([, { matcher }]) => matcher({ labels: testResult.labels }));
                if (!match) {
                    return DEFAULT_ENVIRONMENT_IDENTITY;
                }
                const [id, descriptor] = match;
                return { id, name: descriptor.name ?? id };
            })();
        testResult.environment = environmentIdentity.name;
        __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addEnvironments).call(this, [environmentIdentity]);
        const trHistory = await this.historyByTr(testResult);
        if (trHistory !== undefined) {
            testResult.transition = getStatusTransition(testResult, trHistory);
            testResult.flaky = isFlaky(testResult, trHistory);
        }
        __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").set(testResult.id, testResult);
        __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_setTestResultEnvironmentId).call(this, testResult, environmentIdentity.id);
        hidePreviousAttemptByEnvironment(this.indexLatestEnvTestResultByHistoryId, testResult, environmentIdentity.id);
        index(this.indexTestResultByTestCase, testResult.testCase?.id, testResult);
        index(this.indexTestResultByHistoryId, testResult.historyId, testResult);
        index(this.indexAttachmentByTestResult, testResult.id, ...attachmentLinks);
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeDispatcher, "f")?.sendTestResult(testResult.id);
    }
    async visitTestFixtureResult(result, context) {
        const attachmentLinks = [];
        const testFixtureResult = testFixtureResultRawToState({
            attachments: __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f"),
            visitAttachmentLink: (link) => attachmentLinks.push(link),
        }, result, context);
        __classPrivateFieldGet(this, _DefaultAllureStore_fixtures, "f").set(testFixtureResult.id, testFixtureResult);
        testFixtureResult.testResultIds.forEach((trId) => {
            index(this.indexFixturesByTestResult, trId, testFixtureResult);
        });
        index(this.indexAttachmentByFixture, testFixtureResult.id, ...attachmentLinks);
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeDispatcher, "f")?.sendTestFixtureResult(testFixtureResult.id);
    }
    async visitAttachmentFile(resultFile) {
        const originalFileName = resultFile.getOriginalFileName();
        const id = md5(originalFileName);
        __classPrivateFieldGet(this, _DefaultAllureStore_attachmentContents, "f").set(id, resultFile);
        const maybeLink = __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(id);
        if (maybeLink) {
            const link = maybeLink;
            link.missed = false;
            link.ext = link.ext === undefined || link.ext === "" ? resultFile.getExtension() : link.ext;
            link.contentType = link.contentType ?? resultFile.getContentType();
            link.contentLength = resultFile.getContentLength();
        }
        else {
            __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").set(id, {
                used: false,
                missed: false,
                id,
                originalFileName,
                ext: resultFile.getExtension(),
                contentType: resultFile.getContentType(),
                contentLength: resultFile.getContentLength(),
            });
        }
        for (const globalAttachmentId of __classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIds, "f")) {
            const globalAttachment = __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(globalAttachmentId);
            if (!globalAttachment || globalAttachment.originalFileName !== originalFileName) {
                continue;
            }
            const linkedGlobalAttachment = globalAttachment;
            linkedGlobalAttachment.missed = false;
            linkedGlobalAttachment.ext =
                linkedGlobalAttachment.ext === undefined || linkedGlobalAttachment.ext === ""
                    ? resultFile.getExtension()
                    : linkedGlobalAttachment.ext;
            linkedGlobalAttachment.contentType = linkedGlobalAttachment.contentType ?? resultFile.getContentType();
            linkedGlobalAttachment.contentLength = resultFile.getContentLength();
            __classPrivateFieldGet(this, _DefaultAllureStore_attachmentContents, "f").set(globalAttachmentId, resultFile);
            __classPrivateFieldGet(this, _DefaultAllureStore_realtimeDispatcher, "f")?.sendAttachmentFile(globalAttachmentId);
        }
        __classPrivateFieldGet(this, _DefaultAllureStore_realtimeDispatcher, "f")?.sendAttachmentFile(id);
    }
    async visitMetadata(metadata) {
        Object.keys(metadata).forEach((key) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_metadata, "f").set(key, metadata[key]);
        });
    }
    async visitGlobals(globals) {
        const { errors, attachments } = globals;
        errors.forEach((error) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addGlobalError).call(this, error);
        });
        attachments.forEach((attachment) => {
            const originalFileName = attachment.originalFileName;
            const resolvedEnvironment = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_resolveGlobalEnvironmentIdentity).call(this, attachment.environment);
            const linkedAttachment = __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(md5(originalFileName));
            const attachmentContent = __classPrivateFieldGet(this, _DefaultAllureStore_attachmentContents, "f").get(md5(originalFileName));
            const attachmentLink = {
                id: __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_globalAttachmentId).call(this, originalFileName, resolvedEnvironment?.id),
                name: attachment?.name || originalFileName,
                originalFileName,
                ext: extname(originalFileName),
                used: true,
                missed: !attachmentContent,
                contentType: attachment?.contentType ?? linkedAttachment?.contentType,
                contentLength: linkedAttachment?.contentLength,
                environment: resolvedEnvironment?.name ?? attachment.environment,
            };
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addGlobalAttachment).call(this, attachmentLink, attachmentContent);
        });
    }
    async allTestCases() {
        return Array.from(__classPrivateFieldGet(this, _DefaultAllureStore_testCases, "f").values());
    }
    async allTestResults(options = { includeHidden: false }) {
        const { includeHidden = false, filter } = options;
        const result = [];
        for (const [, tr] of __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")) {
            if (!includeHidden && tr.hidden) {
                continue;
            }
            if (typeof filter === "function" && !filter(tr)) {
                continue;
            }
            result.push(tr);
        }
        return result;
    }
    async allAttachments(options = {}) {
        const { includeMissed = false, includeUnused = false } = options;
        const filteredAttachments = [];
        for (const [, attachment] of __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f")) {
            if (!includeMissed && attachment.missed) {
                continue;
            }
            if (!includeUnused && !attachment.used) {
                continue;
            }
            filteredAttachments.push(attachment);
        }
        return filteredAttachments;
    }
    async allMetadata() {
        const result = {};
        __classPrivateFieldGet(this, _DefaultAllureStore_metadata, "f").forEach((value, key) => (result[key] = value));
        return result;
    }
    async allFixtures() {
        return Array.from(__classPrivateFieldGet(this, _DefaultAllureStore_fixtures, "f").values());
    }
    async allHistoryDataPoints() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_historyPoints, "f");
    }
    async allHistoryDataPointsByEnvironment(environment) {
        const environmentNameValidation = validateEnvironmentName(environment);
        if (!environmentNameValidation.valid) {
            return [];
        }
        const environmentId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, environmentNameValidation.normalized);
        if (!environmentId) {
            return [];
        }
        return this.allHistoryDataPointsByEnvironmentId(environmentId);
    }
    async allHistoryDataPointsByEnvironmentId(environmentId) {
        const normalizedEnvironmentId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdForLookup).call(this, environmentId);
        if (!normalizedEnvironmentId) {
            const validation = validateEnvironmentId(environmentId);
            const reason = validation.valid ? "unknown environment id" : validation.reason;
            throw new Error(`Invalid environmentId ${JSON.stringify(environmentId)}: ${reason}`);
        }
        return __classPrivateFieldGet(this, _DefaultAllureStore_historyPoints, "f").reduce((result, dp) => {
            const filteredTestResults = [];
            for (const tr of Object.values(dp.testResults)) {
                const storedEnvironmentKey = typeof tr.environment === "string" ? tr.environment : undefined;
                const trEnvId = (storedEnvironmentKey ? __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, storedEnvironmentKey) : undefined) ??
                    resolveStoredEnvironmentIdentity({
                        environment: tr.environment,
                        labels: tr.labels ?? [],
                    }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
                        forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
                    })?.id;
                if (trEnvId === normalizedEnvironmentId) {
                    filteredTestResults.push(tr);
                }
            }
            const hasNoEnvironmentTestResults = filteredTestResults.length === 0;
            result.push({
                ...dp,
                testResults: hasNoEnvironmentTestResults
                    ? {}
                    : filteredTestResults.reduce((acc, tr) => {
                        acc[tr.historyId] = tr;
                        return acc;
                    }, {}),
                knownTestCaseIds: hasNoEnvironmentTestResults ? [] : filteredTestResults.map((tr) => tr.id),
            });
            return result;
        }, []);
    }
    async allKnownIssues() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_known, "f");
    }
    async allNewTestResults(filter, history) {
        if (!__classPrivateFieldGet(this, _DefaultAllureStore_history, "f") && !history) {
            return [];
        }
        const allHistoryDps = history ?? (await this.allHistoryDataPoints());
        if (allHistoryDps.length === 0) {
            return Array.from(__classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").values());
        }
        const historicalIds = new Set(allHistoryDps.flatMap((dp) => Object.keys(dp.testResults)));
        const newTrs = [];
        for (const [, tr] of __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")) {
            if (tr.hidden) {
                continue;
            }
            if (typeof filter === "function" && !filter(tr)) {
                continue;
            }
            const historyIdCandidates = getHistoryIdCandidates(tr);
            if (historyIdCandidates.length === 0 || historyIdCandidates.every((historyId) => !historicalIds.has(historyId))) {
                newTrs.push(tr);
            }
        }
        return newTrs;
    }
    async testCaseById(tcId) {
        return __classPrivateFieldGet(this, _DefaultAllureStore_testCases, "f").get(tcId);
    }
    async testResultById(trId) {
        return __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(trId);
    }
    async attachmentById(attachmentId) {
        return __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(attachmentId);
    }
    async attachmentContentById(attachmentId) {
        return __classPrivateFieldGet(this, _DefaultAllureStore_attachmentContents, "f").get(attachmentId);
    }
    async metadataByKey(key) {
        return __classPrivateFieldGet(this, _DefaultAllureStore_metadata, "f").get(key);
    }
    async testResultsByTcId(tcId) {
        return this.indexTestResultByTestCase.get(tcId) ?? [];
    }
    async environmentIdByTrId(trId) {
        const testResult = __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(trId);
        return testResult ? __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByTestResult).call(this, testResult) : undefined;
    }
    async attachmentsByTrId(trId) {
        return this.indexAttachmentByTestResult.get(trId) ?? [];
    }
    async retriesByTr(tr) {
        const retries = [];
        if (!tr || tr.hidden || !tr.historyId) {
            return retries;
        }
        const trByHistoryId = this.indexTestResultByHistoryId.get(tr.historyId);
        if (!trByHistoryId) {
            return retries;
        }
        for (const r of trByHistoryId) {
            if (!r.hidden || tr.environment !== r.environment) {
                continue;
            }
            retries.push(r);
        }
        return retries.sort(nullsLast(compareBy("start", reverse(ordinal()))));
    }
    async retriesByTrId(trId) {
        const tr = await this.testResultById(trId);
        return this.retriesByTr(tr);
    }
    async historyByTr(tr) {
        if (!__classPrivateFieldGet(this, _DefaultAllureStore_history, "f")) {
            return undefined;
        }
        const historyIdCandidates = getHistoryIdCandidates(tr);
        if (historyIdCandidates.length === 0) {
            return [];
        }
        return selectHistoryTestResults(__classPrivateFieldGet(this, _DefaultAllureStore_historyPoints, "f"), historyIdCandidates);
    }
    async historyByTrId(trId) {
        const tr = await this.testResultById(trId);
        if (!tr) {
            return undefined;
        }
        return this.historyByTr(tr);
    }
    async fixturesByTrId(trId) {
        return this.indexFixturesByTestResult.get(trId) ?? [];
    }
    async failedTestResults() {
        const failedTrs = [];
        for (const [, tr] of __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")) {
            if (tr.hidden) {
                continue;
            }
            if (tr.status === "failed" || tr.status === "broken") {
                failedTrs.push(tr);
            }
        }
        return failedTrs;
    }
    async unknownFailedTestResults() {
        const failedTestResults = await this.failedTestResults();
        if (!__classPrivateFieldGet(this, _DefaultAllureStore_known, "f")?.length) {
            return failedTestResults;
        }
        const knownHistoryIds = new Set(__classPrivateFieldGet(this, _DefaultAllureStore_known, "f").map((ktf) => ktf.historyId));
        return failedTestResults.filter((tr) => {
            const historyIdCandidates = getHistoryIdCandidates(tr);
            return (historyIdCandidates.length > 0 && historyIdCandidates.every((historyId) => !knownHistoryIds.has(historyId)));
        });
    }
    async testResultsByLabel(labelName) {
        const results = createDictionary();
        results._ = [];
        for (const [, test] of __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")) {
            if (test.hidden) {
                continue;
            }
            const targetLabels = (test.labels ?? []).filter((label) => label.name === labelName);
            if (targetLabels.length === 0) {
                results._.push(test);
                continue;
            }
            targetLabels.forEach((label) => {
                if (!results[label.value]) {
                    results[label.value] = [];
                }
                results[label.value].push(test);
            });
        }
        return results;
    }
    async testsStatistic(filter) {
        const statistic = { total: 0 };
        for (const [, tr] of __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")) {
            if (tr.hidden) {
                continue;
            }
            if (filter && !filter(tr)) {
                continue;
            }
            statistic.total++;
            const retries = await this.retriesByTr(tr);
            if (retries.length > 0) {
                statistic.retries = (statistic.retries ?? 0) + 1;
            }
            if (tr.flaky) {
                statistic.flaky = (statistic.flaky ?? 0) + 1;
            }
            if (tr.transition === "new") {
                statistic.new = (statistic.new ?? 0) + 1;
            }
            if (!statistic[tr.status]) {
                statistic[tr.status] = 0;
            }
            statistic[tr.status]++;
        }
        return statistic;
    }
    async allEnvironments() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").map(({ name }) => name);
    }
    async allEnvironmentIdentities() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f");
    }
    async testResultsByEnvironment(env, options = { includeHidden: false }) {
        const environmentNameValidation = validateEnvironmentName(env);
        if (!environmentNameValidation.valid) {
            return [];
        }
        const environmentId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, environmentNameValidation.normalized);
        if (!environmentId) {
            return [];
        }
        return this.testResultsByEnvironmentId(environmentId, options);
    }
    async testResultsByEnvironmentId(envId, options = { includeHidden: false }) {
        const normalizedEnvId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdForLookup).call(this, envId);
        if (!normalizedEnvId) {
            const validation = validateEnvironmentId(envId);
            const reason = validation.valid ? "unknown environment id" : validation.reason;
            throw new Error(`Invalid environmentId ${JSON.stringify(envId)}: ${reason}`);
        }
        const trs = this.indexTestResultByEnvironmentId.get(normalizedEnvId) ?? [];
        return options.includeHidden ? [...trs] : trs.filter((tr) => !tr.hidden);
    }
    async allTestEnvGroups() {
        const trByTestCaseId = {};
        for (const [, tr] of __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")) {
            const testCaseId = tr?.testCase?.id;
            if (!testCaseId) {
                continue;
            }
            if (trByTestCaseId[testCaseId]) {
                trByTestCaseId[testCaseId].push(tr);
            }
            else {
                trByTestCaseId[testCaseId] = [tr];
            }
        }
        return Object.entries(trByTestCaseId).reduce((acc, [testCaseId, trs]) => {
            if (trs.length === 0) {
                return acc;
            }
            const { fullName, name } = trs[0];
            const envGroup = {
                id: testCaseId,
                fullName,
                name,
                status: getWorstStatus(trs.map(({ status }) => status)) ?? "passed",
                testResultsByEnv: {},
            };
            trs.forEach((tr) => {
                const env = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByTestResult).call(this, tr) ?? DEFAULT_ENVIRONMENT;
                envGroup.testResultsByEnv[env] = tr.id;
            });
            acc.push(envGroup);
            return acc;
        }, []);
    }
    async allVariables() {
        return __classPrivateFieldGet(this, _DefaultAllureStore_reportVariables, "f");
    }
    async envVariables(env) {
        const environmentNameValidation = validateEnvironmentName(env);
        if (!environmentNameValidation.valid) {
            return {
                ...__classPrivateFieldGet(this, _DefaultAllureStore_reportVariables, "f"),
            };
        }
        const environmentId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, environmentNameValidation.normalized);
        if (!environmentId) {
            return {
                ...__classPrivateFieldGet(this, _DefaultAllureStore_reportVariables, "f"),
            };
        }
        return this.envVariablesByEnvironmentId(environmentId);
    }
    async envVariablesByEnvironmentId(envId) {
        const normalizedEnvId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdForLookup).call(this, envId);
        if (!normalizedEnvId) {
            const validation = validateEnvironmentId(envId);
            const reason = validation.valid ? "unknown environment id" : validation.reason;
            throw new Error(`Invalid environmentId ${JSON.stringify(envId)}: ${reason}`);
        }
        const envDescriptor = Object.prototype.hasOwnProperty.call(__classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), normalizedEnvId)
            ? __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f")[normalizedEnvId]
            : undefined;
        return {
            ...__classPrivateFieldGet(this, _DefaultAllureStore_reportVariables, "f"),
            ...(envDescriptor?.variables ?? {}),
        };
    }
    dumpState() {
        const storeDump = {
            testResults: mapToObject(__classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f")),
            attachments: mapToObject(__classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f")),
            testCases: mapToObject(__classPrivateFieldGet(this, _DefaultAllureStore_testCases, "f")),
            fixtures: mapToObject(__classPrivateFieldGet(this, _DefaultAllureStore_fixtures, "f")),
            environments: __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f"),
            reportVariables: __classPrivateFieldGet(this, _DefaultAllureStore_reportVariables, "f"),
            globalAttachmentIds: __classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIds, "f"),
            globalErrors: __classPrivateFieldGet(this, _DefaultAllureStore_globalErrors, "f"),
            indexLatestEnvTestResultByHistoryId: {},
            indexAttachmentByTestResult: {},
            indexTestResultByHistoryId: {},
            indexTestResultByTestCase: {},
            indexAttachmentByFixture: {},
            indexFixturesByTestResult: {},
            indexKnownByHistoryId: {},
            qualityGateResults: __classPrivateFieldGet(this, _DefaultAllureStore_qualityGateResults, "f"),
        };
        this.indexLatestEnvTestResultByHistoryId.forEach((envMap, environmentId) => {
            if (!storeDump.indexLatestEnvTestResultByHistoryId[environmentId]) {
                storeDump.indexLatestEnvTestResultByHistoryId[environmentId] = {};
            }
            envMap.forEach((tr, historyId) => {
                storeDump.indexLatestEnvTestResultByHistoryId[environmentId][historyId] = tr.id;
            });
        });
        this.indexAttachmentByFixture.forEach((link, fxId) => {
            storeDump.indexAttachmentByFixture[fxId] = link.map((l) => l.id);
        });
        this.indexAttachmentByTestResult.forEach((links, trId) => {
            storeDump.indexAttachmentByTestResult[trId] = links.map((l) => l.id);
        });
        this.indexTestResultByHistoryId.forEach((trs, historyId) => {
            storeDump.indexTestResultByHistoryId[historyId] = trs.map((tr) => tr.id);
        });
        this.indexTestResultByTestCase.forEach((trs, tcId) => {
            storeDump.indexTestResultByTestCase[tcId] = trs.map((tr) => tr.id);
        });
        this.indexFixturesByTestResult.forEach((fixtures, trId) => {
            storeDump.indexFixturesByTestResult[trId] = fixtures.map((f) => f.id);
        });
        this.indexKnownByHistoryId.forEach((known, historyId) => {
            storeDump.indexKnownByHistoryId[historyId] = known;
        });
        return storeDump;
    }
    async restoreState(stateDump, attachmentsContents = {}) {
        const { testResults, attachments, testCases, fixtures, reportVariables, environments, globalAttachmentIds = [], globalErrors = [], indexAttachmentByTestResult = {}, indexTestResultByHistoryId = {}, indexTestResultByTestCase = {}, indexLatestEnvTestResultByHistoryId = {}, indexAttachmentByFixture = {}, indexFixturesByTestResult = {}, indexKnownByHistoryId = {}, qualityGateResults = [], } = stateDump;
        const storedEnvironmentAliases = environments.flatMap((environmentValue) => {
            if (typeof environmentValue === "string") {
                return [{ id: environmentValue, name: environmentValue }];
            }
            const idValidation = validateEnvironmentId(environmentValue.id);
            if (!idValidation.valid) {
                return [];
            }
            return [
                {
                    id: idValidation.normalized,
                    name: environmentValue.name ?? idValidation.normalized,
                },
            ];
        });
        const normalizedEnvironments = environments
            .map((environmentValue) => {
            if (typeof environmentValue === "string") {
                return resolveStoredEnvironmentIdentity({
                    environmentName: environmentValue,
                }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
                    forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
                    fallbackToMatch: false,
                });
            }
            return resolveStoredEnvironmentIdentity({
                environment: environmentValue.id,
                environmentName: environmentValue.name,
            }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
                forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
                fallbackToMatch: false,
            });
        })
            .filter(Boolean);
        environments.forEach((environmentValue, index) => {
            const environmentParams = typeof environmentValue === "string"
                ? { environmentName: environmentValue }
                : { environment: environmentValue.id, environmentName: environmentValue.name };
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedStoredEnvironment).call(this, environmentParams, `restored environments[${index}]`, {
                fallbackToMatch: false,
            });
        });
        __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addEnvironments).call(this, [...storedEnvironmentAliases, ...normalizedEnvironments]);
        const envNameToId = new Map();
        for (const { id, name } of __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f")) {
            envNameToId.set(name, id);
            envNameToId.set(id, id);
        }
        Object.values(testResults).forEach((testResult) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").set(testResult.id, testResult);
            const storedEnvKey = typeof testResult.environment === "string" ? testResult.environment : undefined;
            const envId = (storedEnvKey ? envNameToId.get(storedEnvKey) : undefined) ?? __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByTestResult).call(this, testResult);
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedEnvironmentId).call(this, envId, `restored testResults[${JSON.stringify(testResult.id)}]`);
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_setTestResultEnvironmentId).call(this, testResult, envId);
        });
        updateMapWithRecord(__classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f"), attachments);
        updateMapWithRecord(__classPrivateFieldGet(this, _DefaultAllureStore_testCases, "f"), testCases);
        updateMapWithRecord(__classPrivateFieldGet(this, _DefaultAllureStore_fixtures, "f"), fixtures);
        updateMapWithRecord(__classPrivateFieldGet(this, _DefaultAllureStore_attachmentContents, "f"), attachmentsContents);
        globalAttachmentIds.forEach((id) => {
            const attachment = __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(id);
            if (!attachment) {
                return;
            }
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedStoredEnvironment).call(this, {
                environment: attachment.environment,
                environmentName: attachment.environment,
            }, `restored globalAttachments[${JSON.stringify(id)}]`, {
                fallbackToMatch: false,
            });
            __classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIds, "f").push(id);
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_indexGlobalAttachment).call(this, attachment);
        });
        globalErrors.forEach((error, index) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedStoredEnvironment).call(this, {
                environment: error.environment,
                environmentName: error.environment,
            }, `restored globalErrors[${index}]`, {
                fallbackToMatch: false,
            });
            __classPrivateFieldGet(this, _DefaultAllureStore_globalErrors, "f").push(__classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_indexGlobalError).call(this, error));
        });
        Object.assign(__classPrivateFieldGet(this, _DefaultAllureStore_reportVariables, "f"), reportVariables);
        Object.entries(indexAttachmentByTestResult).forEach(([trId, links]) => {
            const attachmentsLinks = links.map((id) => __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(id)).filter(Boolean);
            if (attachmentsLinks.length === 0) {
                return;
            }
            const existingLinks = this.indexAttachmentByTestResult.get(trId);
            if (!existingLinks) {
                this.indexAttachmentByTestResult.set(trId, attachmentsLinks);
                return;
            }
            existingLinks.push(...attachmentsLinks);
        });
        Object.entries(indexTestResultByHistoryId).forEach(([historyId, trIds]) => {
            const trs = trIds.map((id) => __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(id)).filter(Boolean);
            if (trs.length === 0) {
                return;
            }
            const existingTrs = this.indexTestResultByHistoryId.get(historyId);
            if (!existingTrs) {
                this.indexTestResultByHistoryId.set(historyId, trs);
                return;
            }
            existingTrs.push(...trs);
        });
        Object.entries(indexTestResultByTestCase).forEach(([tcId, trIds]) => {
            const trs = trIds.map((id) => __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(id)).filter(Boolean);
            if (trs.length === 0) {
                return;
            }
            const existingTrs = this.indexTestResultByTestCase.get(tcId);
            if (!existingTrs) {
                this.indexTestResultByTestCase.set(tcId, trs);
                return;
            }
            existingTrs.push(...trs);
        });
        Object.entries(indexAttachmentByFixture).forEach(([fxId, attachmentIds]) => {
            const attachmentsLinks = attachmentIds.map((id) => __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").get(id)).filter(Boolean);
            if (attachmentsLinks.length === 0) {
                return;
            }
            const existingLinks = this.indexAttachmentByFixture.get(fxId);
            if (!existingLinks) {
                this.indexAttachmentByFixture.set(fxId, attachmentsLinks);
                return;
            }
            existingLinks.push(...attachmentsLinks);
        });
        Object.entries(indexFixturesByTestResult).forEach(([trId, fixtureIds]) => {
            const fxs = fixtureIds.map((id) => __classPrivateFieldGet(this, _DefaultAllureStore_fixtures, "f").get(id)).filter(Boolean);
            if (fxs.length === 0) {
                return;
            }
            const existingFixtures = this.indexFixturesByTestResult.get(trId);
            if (!existingFixtures) {
                this.indexFixturesByTestResult.set(trId, fxs);
                return;
            }
            existingFixtures.push(...fxs);
        });
        Object.entries(indexKnownByHistoryId).forEach(([historyId, knownFailures]) => {
            const existingKnown = this.indexKnownByHistoryId.get(historyId);
            if (!existingKnown) {
                this.indexKnownByHistoryId.set(historyId, knownFailures);
                return;
            }
            existingKnown.push(...knownFailures);
        });
        __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").forEach(({ id }) => {
            if (!this.indexLatestEnvTestResultByHistoryId.has(id)) {
                this.indexLatestEnvTestResultByHistoryId.set(id, new Map());
            }
        });
        const latestAttemptsEntries = Object.entries(indexLatestEnvTestResultByHistoryId);
        const hasNestedLatestAttempts = latestAttemptsEntries.some(([, value]) => typeof value === "object" && value !== null);
        if (hasNestedLatestAttempts) {
            latestAttemptsEntries.forEach(([environmentId, historyIds]) => {
                const environmentIdValidation = validateEnvironmentId(environmentId);
                if (!environmentIdValidation.valid || typeof historyIds !== "object" || historyIds === null) {
                    return;
                }
                const normalizedEnvironmentId = environmentIdValidation.normalized;
                __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedEnvironmentId).call(this, normalizedEnvironmentId, `restored indexLatestEnvTestResultByHistoryId[${JSON.stringify(environmentId)}]`);
                Object.values(historyIds).forEach((trId) => {
                    const tr = __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(trId);
                    if (!tr) {
                        return;
                    }
                    hidePreviousAttemptByEnvironment(this.indexLatestEnvTestResultByHistoryId, tr, normalizedEnvironmentId);
                });
            });
        }
        else {
            Object.entries(indexLatestEnvTestResultByHistoryId).forEach(([historyId, latestTrId]) => {
                const trIdsForHistory = indexTestResultByHistoryId[historyId] ?? [];
                const storedLatestTr = __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(latestTrId);
                const storedLatestEnvironmentId = storedLatestTr
                    ? (__classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByTestResult).call(this, storedLatestTr) ?? DEFAULT_ENVIRONMENT)
                    : undefined;
                if (storedLatestTr && storedLatestEnvironmentId) {
                    hidePreviousAttemptByEnvironment(this.indexLatestEnvTestResultByHistoryId, storedLatestTr, storedLatestEnvironmentId);
                }
                for (const trId of trIdsForHistory) {
                    if (trId === latestTrId) {
                        continue;
                    }
                    const tr = __classPrivateFieldGet(this, _DefaultAllureStore_testResults, "f").get(trId);
                    if (!tr) {
                        continue;
                    }
                    const environmentId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByTestResult).call(this, tr) ?? DEFAULT_ENVIRONMENT;
                    if (storedLatestEnvironmentId && environmentId === storedLatestEnvironmentId) {
                        continue;
                    }
                    hidePreviousAttemptByEnvironment(this.indexLatestEnvTestResultByHistoryId, tr, environmentId);
                }
            });
        }
        qualityGateResults.forEach((result, index) => {
            __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedStoredEnvironment).call(this, {
                environment: result.environment,
                environmentName: result.environment,
            }, `restored qualityGateResults[${index}]`, {
                fallbackToMatch: false,
            });
            __classPrivateFieldGet(this, _DefaultAllureStore_qualityGateResults, "f").push(result);
        });
    }
}
_DefaultAllureStore_testResults = new WeakMap(), _DefaultAllureStore_environmentDisplayNames = new WeakMap(), _DefaultAllureStore_environmentNameToId = new WeakMap(), _DefaultAllureStore_attachments = new WeakMap(), _DefaultAllureStore_attachmentContents = new WeakMap(), _DefaultAllureStore_testCases = new WeakMap(), _DefaultAllureStore_metadata = new WeakMap(), _DefaultAllureStore_history = new WeakMap(), _DefaultAllureStore_known = new WeakMap(), _DefaultAllureStore_fixtures = new WeakMap(), _DefaultAllureStore_defaultLabels = new WeakMap(), _DefaultAllureStore_environment = new WeakMap(), _DefaultAllureStore_environmentsConfig = new WeakMap(), _DefaultAllureStore_reportVariables = new WeakMap(), _DefaultAllureStore_realtimeDispatcher = new WeakMap(), _DefaultAllureStore_realtimeSubscriber = new WeakMap(), _DefaultAllureStore_allowedEnvironmentIds = new WeakMap(), _DefaultAllureStore_globalAttachmentIds = new WeakMap(), _DefaultAllureStore_globalAttachmentIdsByEnv = new WeakMap(), _DefaultAllureStore_globalErrors = new WeakMap(), _DefaultAllureStore_globalErrorsByEnv = new WeakMap(), _DefaultAllureStore_globalExitCode = new WeakMap(), _DefaultAllureStore_qualityGateResults = new WeakMap(), _DefaultAllureStore_historyPoints = new WeakMap(), _DefaultAllureStore_environments = new WeakMap(), _DefaultAllureStore_instances = new WeakSet(), _DefaultAllureStore_mergeEnvironmentIdentity = function _DefaultAllureStore_mergeEnvironmentIdentity(existingEnvironment, incomingEnvironment) {
    const configuredEnvironment = environmentIdentityById(__classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), incomingEnvironment.id);
    if (configuredEnvironment) {
        return configuredEnvironment;
    }
    if (existingEnvironment) {
        const existingNameValidation = validateEnvironmentName(existingEnvironment.name);
        if (existingNameValidation.valid) {
            return {
                id: existingEnvironment.id,
                name: existingNameValidation.normalized,
            };
        }
    }
    const incomingNameValidation = validateEnvironmentName(incomingEnvironment.name);
    if (incomingNameValidation.valid) {
        return {
            id: incomingEnvironment.id,
            name: incomingNameValidation.normalized,
        };
    }
    return {
        id: incomingEnvironment.id,
        name: incomingEnvironment.id,
    };
}, _DefaultAllureStore_environmentIdForLookup = function _DefaultAllureStore_environmentIdForLookup(environmentKey) {
    const aliasedEnvironmentId = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, environmentKey);
    if (aliasedEnvironmentId) {
        return aliasedEnvironmentId;
    }
    const environmentIdValidation = validateEnvironmentId(environmentKey);
    return environmentIdValidation.valid ? environmentIdValidation.normalized : undefined;
}, _DefaultAllureStore_addEnvironments = function _DefaultAllureStore_addEnvironments(envs) {
    if (__classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").length === 0) {
        __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").push(DEFAULT_ENVIRONMENT_IDENTITY);
    }
    const nextById = new Map(__classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").map((environment) => [environment.id, environment]));
    envs.forEach((environment) => {
        nextById.set(environment.id, __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_mergeEnvironmentIdentity).call(this, nextById.get(environment.id), environment));
    });
    __classPrivateFieldSet(this, _DefaultAllureStore_environments, Array.from(nextById.values()), "f");
    __classPrivateFieldGet(this, _DefaultAllureStore_environmentNameToId, "f").clear();
    __classPrivateFieldGet(this, _DefaultAllureStore_environments, "f").forEach(({ id, name }) => {
        __classPrivateFieldGet(this, _DefaultAllureStore_environmentNameToId, "f").set(name, id);
    });
    envs.forEach(({ id, name }) => {
        __classPrivateFieldGet(this, _DefaultAllureStore_environmentDisplayNames, "f").set(name, id);
    });
    envs.forEach(({ id }) => {
        if (!this.indexLatestEnvTestResultByHistoryId.has(id)) {
            this.indexLatestEnvTestResultByHistoryId.set(id, new Map());
        }
    });
}, _DefaultAllureStore_environmentIdByName = function _DefaultAllureStore_environmentIdByName(environmentName) {
    const canonicalId = __classPrivateFieldGet(this, _DefaultAllureStore_environmentNameToId, "f").get(environmentName);
    if (canonicalId) {
        return canonicalId;
    }
    return __classPrivateFieldGet(this, _DefaultAllureStore_environmentDisplayNames, "f").get(environmentName);
}, _DefaultAllureStore_setTestResultEnvironmentId = function _DefaultAllureStore_setTestResultEnvironmentId(testResult, environmentId) {
    if (!environmentId) {
        return;
    }
    const existing = this.indexTestResultByEnvironmentId.get(environmentId);
    if (existing?.some((tr) => tr.id === testResult.id)) {
        return;
    }
    index(this.indexTestResultByEnvironmentId, environmentId, testResult);
}, _DefaultAllureStore_assertAllowedEnvironmentId = function _DefaultAllureStore_assertAllowedEnvironmentId(environmentId, sourcePath) {
    if (!environmentId) {
        return;
    }
    const error = validateAllowedEnvironmentId(environmentId, __classPrivateFieldGet(this, _DefaultAllureStore_allowedEnvironmentIds, "f"), sourcePath);
    if (error) {
        throw new Error(error);
    }
}, _DefaultAllureStore_assertAllowedStoredEnvironment = function _DefaultAllureStore_assertAllowedStoredEnvironment(params, sourcePath, options) {
    const identity = resolveStoredEnvironmentIdentity(params, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
        forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
        fallbackToMatch: options?.fallbackToMatch,
    });
    const environmentKey = identity?.id ??
        (typeof params.environment === "string" ? params.environment : undefined) ??
        (typeof params.environmentName === "string" ? params.environmentName : undefined);
    if (environmentKey === undefined) {
        return;
    }
    __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_assertAllowedEnvironmentId).call(this, environmentKey, sourcePath);
}, _DefaultAllureStore_environmentIdByTestResult = function _DefaultAllureStore_environmentIdByTestResult(testResult) {
    const storedEnvironmentKey = typeof testResult.environment === "string" ? testResult.environment : undefined;
    return ((storedEnvironmentKey ? __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_environmentIdByName).call(this, storedEnvironmentKey) : undefined) ??
        resolveStoredEnvironmentIdentity({
            environment: testResult.environment,
            labels: testResult.labels,
        }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
            forcedEnvironment: __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f"),
        })?.id);
}, _DefaultAllureStore_resolveGlobalEnvironmentIdentity = function _DefaultAllureStore_resolveGlobalEnvironmentIdentity(environment) {
    if (environment !== undefined) {
        const resolvedEnvironment = resolveStoredEnvironmentIdentity({
            environment,
            environmentName: environment,
        }, __classPrivateFieldGet(this, _DefaultAllureStore_environmentsConfig, "f"), {
            fallbackToMatch: false,
        });
        if (resolvedEnvironment) {
            return resolvedEnvironment;
        }
    }
    if (__classPrivateFieldGet(this, _DefaultAllureStore_environment, "f")) {
        return __classPrivateFieldGet(this, _DefaultAllureStore_environment, "f");
    }
    return DEFAULT_ENVIRONMENT_IDENTITY;
}, _DefaultAllureStore_globalAttachmentId = function _DefaultAllureStore_globalAttachmentId(originalFileName, environmentId) {
    return md5(environmentId ? `${environmentId}:${originalFileName}` : originalFileName);
}, _DefaultAllureStore_indexGlobalError = function _DefaultAllureStore_indexGlobalError(error) {
    const resolvedEnvironment = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_resolveGlobalEnvironmentIdentity).call(this, error.environment);
    if (!resolvedEnvironment) {
        return error;
    }
    error.environment = resolvedEnvironment.name;
    __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addEnvironments).call(this, [resolvedEnvironment]);
    index(__classPrivateFieldGet(this, _DefaultAllureStore_globalErrorsByEnv, "f"), resolvedEnvironment.id, error);
    return error;
}, _DefaultAllureStore_addGlobalError = function _DefaultAllureStore_addGlobalError(error) {
    __classPrivateFieldGet(this, _DefaultAllureStore_globalErrors, "f").push(__classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_indexGlobalError).call(this, error));
}, _DefaultAllureStore_indexGlobalAttachment = function _DefaultAllureStore_indexGlobalAttachment(attachmentLink) {
    const resolvedEnvironment = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_resolveGlobalEnvironmentIdentity).call(this, attachmentLink.environment);
    if (!resolvedEnvironment) {
        return attachmentLink;
    }
    attachmentLink.environment = resolvedEnvironment.name;
    __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_addEnvironments).call(this, [resolvedEnvironment]);
    index(__classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIdsByEnv, "f"), resolvedEnvironment.id, attachmentLink.id);
    return attachmentLink;
}, _DefaultAllureStore_addGlobalAttachment = function _DefaultAllureStore_addGlobalAttachment(attachmentLink, attachment) {
    const indexedAttachment = __classPrivateFieldGet(this, _DefaultAllureStore_instances, "m", _DefaultAllureStore_indexGlobalAttachment).call(this, { ...attachmentLink });
    __classPrivateFieldGet(this, _DefaultAllureStore_attachments, "f").set(indexedAttachment.id, indexedAttachment);
    if (attachment) {
        __classPrivateFieldGet(this, _DefaultAllureStore_attachmentContents, "f").set(indexedAttachment.id, attachment);
    }
    __classPrivateFieldGet(this, _DefaultAllureStore_globalAttachmentIds, "f").push(indexedAttachment.id);
};
