// @allurereport/plugin-awesome/dist/plugin.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _AwesomePlugin_writer, _AwesomePlugin_generate;
import { incrementStatistic, joinPosixPath } from "@allurereport/core-api";
import { createPluginSummary, } from "@allurereport/plugin-api";
import { preciseTreeLabels } from "@allurereport/plugin-api";
import { applyCategoriesToTestResults, generateCategories } from "./categories.js";
import { generateTimeline } from "./generateTimeline.js";
import { generateAllCharts, generateAttachmentsFiles, generateEnvironmentJson, generateEnvirontmentsList, generateGlobals, generateHistoryDataPoints, generateNav, generateQualityGateResults, generateStaticFiles, generateStatistic, generateTestCases, generateTestEnvGroups, generateTestResults, generateTree, generateTreeFilters, generateVariables, } from "./generators.js";
import { InMemoryReportDataWriter, ReportFileDataWriter } from "./writer.js";
const statisticByTestResults = async (store, testResults) => {
    const statistic = { total: 0 };
    for (const testResult of testResults) {
        if (testResult.hidden) {
            continue;
        }
        statistic.total++;
        incrementStatistic(statistic, testResult.status);
        if ((await store.retriesByTrId(testResult.id)).length > 0) {
            statistic.retries = (statistic.retries ?? 0) + 1;
        }
        if (testResult.flaky) {
            statistic.flaky = (statistic.flaky ?? 0) + 1;
        }
        if (testResult.transition === "new") {
            statistic.new = (statistic.new ?? 0) + 1;
        }
    }
    return statistic;
};
export class AwesomePlugin {
    constructor(options = {}) {
        this.options = options;
        _AwesomePlugin_writer.set(this, void 0);
        _AwesomePlugin_generate.set(this, async (context, store) => {
            const { singleFile, groupBy = [], filter, appendTitlePath } = this.options ?? {};
            const hideLabels = context.hideLabels;
            const categories = context.categories ?? [];
            const environmentItems = await store.metadataByKey("allure_environment");
            const attachments = await store.allAttachments();
            const allTrs = await store.allTestResults({ includeHidden: true, filter });
            const statistics = await store.testsStatistic(filter);
            const environments = await store.allEnvironmentIdentities();
            const envStatistics = new Map();
            const allTestEnvGroups = await store.allTestEnvGroups();
            const globalAttachments = await store.allGlobalAttachments();
            const globalAttachmentsByEnv = await store.allGlobalAttachmentsByEnv();
            const globalExitCode = await store.globalExitCode();
            const globalErrors = await store.allGlobalErrors();
            const globalErrorsByEnv = await store.allGlobalErrorsByEnv();
            const qualityGateResults = await store.qualityGateResultsByEnvironmentId();
            const envResultsById = new Map();
            const envIdByTrId = new Map();
            environments.forEach(({ id }) => {
                envResultsById.set(id, []);
            });
            await Promise.all(allTrs.map(async (tr) => {
                const environmentId = await store.environmentIdByTrId(tr.id);
                if (!environmentId) {
                    return;
                }
                envIdByTrId.set(tr.id, environmentId);
                if (!envResultsById.has(environmentId)) {
                    envResultsById.set(environmentId, []);
                }
                envResultsById.get(environmentId).push(tr);
            }));
            await Promise.all(environments.map(async ({ id }) => {
                envStatistics.set(id, await statisticByTestResults(store, envResultsById.get(id) ?? []));
            }));
            await generateStatistic(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                stats: statistics,
                statsByEnv: envStatistics,
                envs: environments,
            });
            await generateAllCharts(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store, this.options, context);
            const convertedTrs = await generateTestResults(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store, allTrs, { hideLabels });
            applyCategoriesToTestResults(convertedTrs, categories);
            await generateCategories(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                tests: convertedTrs,
                categories,
                environmentCount: environments.length,
                environments: environments.map(({ name }) => name),
                defaultEnvironment: "default",
                selectedEnvironmentCount: environments.length,
            });
            const hasGroupBy = groupBy.length > 0;
            await generateTimeline(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), allTrs, this.options, envIdByTrId);
            const treeLabels = hasGroupBy
                ? preciseTreeLabels(groupBy, convertedTrs, ({ labels }) => labels.map(({ name }) => name))
                : [];
            await generateHistoryDataPoints(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store);
            await generateTestCases(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs);
            await generateTree(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), "tree.json", treeLabels, convertedTrs, { appendTitlePath });
            await generateNav(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs, "nav.json");
            await generateTestEnvGroups(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), allTestEnvGroups);
            const envConvertedTrsById = new Map();
            convertedTrs.forEach((tr) => {
                const environmentId = envIdByTrId.get(tr.id);
                if (!environmentId) {
                    return;
                }
                if (!envConvertedTrsById.has(environmentId)) {
                    envConvertedTrsById.set(environmentId, []);
                }
                envConvertedTrsById.get(environmentId).push(tr);
            });
            for (const reportEnvironment of environments) {
                const envConvertedTrs = envConvertedTrsById.get(reportEnvironment.id) ?? [];
                await generateTree(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), joinPosixPath(reportEnvironment.id, "tree.json"), treeLabels, envConvertedTrs, {
                    appendTitlePath,
                });
                await generateNav(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), envConvertedTrs, joinPosixPath(reportEnvironment.id, "nav.json"));
                await generateCategories(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                    tests: envConvertedTrs,
                    categories,
                    environmentCount: 1,
                    defaultEnvironment: "default",
                    selectedEnvironmentCount: 1,
                    filename: joinPosixPath(reportEnvironment.id, "categories.json"),
                });
            }
            await generateTreeFilters(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), convertedTrs);
            await generateEnvirontmentsList(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store);
            await generateVariables(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), store);
            await generateEnvironmentJson(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), environmentItems ?? []);
            if (attachments?.length) {
                await generateAttachmentsFiles(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), attachments, (id) => store.attachmentContentById(id));
            }
            await generateQualityGateResults(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), qualityGateResults);
            await generateGlobals(__classPrivateFieldGet(this, _AwesomePlugin_writer, "f"), {
                globalAttachments,
                globalAttachmentsByEnv,
                globalErrors,
                globalErrorsByEnv,
                globalExitCode,
                contentFunction: (id) => store.attachmentContentById(id),
            });
            const reportDataFiles = singleFile ? __classPrivateFieldGet(this, _AwesomePlugin_writer, "f").reportFiles() : [];
            await generateStaticFiles({
                ...this.options,
                id: context.id,
                allureVersion: context.allureVersion,
                reportFiles: context.reportFiles,
                reportUuid: context.reportUuid,
                reportName: context.reportName,
                ci: context.ci,
                reportDataFiles,
            });
        });
        this.start = async (context) => {
            const { singleFile } = this.options;
            if (singleFile) {
                __classPrivateFieldSet(this, _AwesomePlugin_writer, new InMemoryReportDataWriter(), "f");
                return;
            }
            __classPrivateFieldSet(this, _AwesomePlugin_writer, new ReportFileDataWriter(context.reportFiles), "f");
            await Promise.resolve();
        };
        this.update = async (context, store) => {
            if (!__classPrivateFieldGet(this, _AwesomePlugin_writer, "f")) {
                throw new Error("call start first");
            }
            await __classPrivateFieldGet(this, _AwesomePlugin_generate, "f").call(this, context, store);
        };
        this.done = async (context, store) => {
            if (!__classPrivateFieldGet(this, _AwesomePlugin_writer, "f")) {
                throw new Error("call start first");
            }
            await __classPrivateFieldGet(this, _AwesomePlugin_generate, "f").call(this, context, store);
        };
    }
    async info(context, store) {
        return createPluginSummary({
            name: this.options.reportName || context.reportName,
            plugin: "Awesome",
            meta: {
                reportId: context.reportUuid,
                singleFile: this.options.singleFile ?? false,
                withTestResultsLinks: true,
            },
            filter: this.options.filter,
            ci: context.ci,
            history: context.history,
            store,
        });
    }
}
_AwesomePlugin_writer = new WeakMap(), _AwesomePlugin_generate = new WeakMap();
