// @allurereport/plugin-allure2/dist/writer.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _InMemoryReportDataWriter_data;
import { mkdir, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { joinPosixPath } from "@allurereport/core-api";
export class FileSystemReportDataWriter {
    constructor(output) {
        this.output = output;
    }
    async writeData(fileName, data) {
        const distFolder = resolve(this.output, "data");
        await mkdir(distFolder, { recursive: true });
        await writeFile(resolve(distFolder, fileName), JSON.stringify(data), { encoding: "utf-8" });
    }
    async writeWidget(fileName, data) {
        const distFolder = resolve(this.output, "widgets");
        await mkdir(distFolder, { recursive: true });
        await writeFile(resolve(distFolder, fileName), JSON.stringify(data), { encoding: "utf-8" });
    }
    async writeTestCase(test) {
        const distFolder = resolve(this.output, "data", "test-cases");
        await mkdir(distFolder, { recursive: true });
        await writeFile(resolve(distFolder, `${test.uid}.json`), JSON.stringify(test), { encoding: "utf-8" });
    }
    async writeAttachment(source, file) {
        const distFolder = resolve(this.output, "data", "attachments");
        await mkdir(distFolder, { recursive: true });
        await file.writeTo(resolve(distFolder, source));
    }
}
export class InMemoryReportDataWriter {
    constructor() {
        _InMemoryReportDataWriter_data.set(this, {});
    }
    async writeData(fileName, data) {
        const dist = joinPosixPath("data", fileName);
        __classPrivateFieldGet(this, _InMemoryReportDataWriter_data, "f")[dist] = Buffer.from(JSON.stringify(data), "utf-8");
    }
    async writeWidget(fileName, data) {
        const dist = joinPosixPath("widgets", fileName);
        __classPrivateFieldGet(this, _InMemoryReportDataWriter_data, "f")[dist] = Buffer.from(JSON.stringify(data), "utf-8");
    }
    async writeTestCase(test) {
        const dist = joinPosixPath("data", "test-cases", `${test.uid}.json`);
        __classPrivateFieldGet(this, _InMemoryReportDataWriter_data, "f")[dist] = Buffer.from(JSON.stringify(test), "utf-8");
    }
    async writeAttachment(fileName, file) {
        const dist = joinPosixPath("data", "attachments", fileName);
        const content = await file.asBuffer();
        if (content) {
            __classPrivateFieldGet(this, _InMemoryReportDataWriter_data, "f")[dist] = content;
        }
    }
    reportFiles() {
        return Object.keys(__classPrivateFieldGet(this, _InMemoryReportDataWriter_data, "f")).map((key) => ({
            name: key,
            value: __classPrivateFieldGet(this, _InMemoryReportDataWriter_data, "f")[key].toString("base64"),
        }));
    }
}
_InMemoryReportDataWriter_data = new WeakMap();
export class ReportFileDataWriter {
    constructor(reportFiles) {
        this.reportFiles = reportFiles;
    }
    async writeData(fileName, data) {
        await this.reportFiles.addFile(joinPosixPath("data", fileName), Buffer.from(JSON.stringify(data), "utf-8"));
    }
    async writeWidget(fileName, data) {
        await this.reportFiles.addFile(joinPosixPath("widgets", fileName), Buffer.from(JSON.stringify(data), "utf-8"));
    }
    async writeAttachment(source, file) {
        const contentBuffer = await file.asBuffer();
        if (!contentBuffer) {
            return;
        }
        await this.reportFiles.addFile(joinPosixPath("data", "attachments", source), contentBuffer);
    }
    async writeTestCase(test) {
        await this.reportFiles.addFile(joinPosixPath("data", "test-cases", `${test.uid}.json`), Buffer.from(JSON.stringify(test), "utf8"));
    }
}
