// allure/dist/utils/process.js
import { spawn } from "node:child_process";
import process from "node:process";
import { platform } from "process";
import { invokeStdoutCliTool } from "@allurereport/reader-api";
const IS_WIN = platform === "win32";
const PS_OUTPUT_PATTERN = /(?<ppid>\d+)\s+(?<pid>\d+)\s+(?<comm>.*)/;
export const runProcess = (params) => {
    const { command, commandArgs, cwd, environmentVariables = {}, logs = "inherit" } = params;
    const env = {
        ...process.env,
        ...environmentVariables,
    };
    if (logs === "pipe") {
        Object.assign(env, {
            FORCE_COLOR: "1",
            CLICOLOR_FORCE: "1",
            COLOR: "1",
            COLORTERM: "truecolor",
            TERM: "xterm-256color",
        });
        delete env.NO_COLOR;
    }
    return spawn(command, commandArgs, {
        env,
        cwd,
        stdio: logs,
        shell: IS_WIN,
    });
};
export const terminationOf = (testProcess) => new Promise((resolve) => {
    testProcess.on("exit", (code) => resolve(code));
});
export const stopProcessTree = async (pid, { signal = "SIGTERM" } = {}) => {
    const tree = new Map();
    const processesToSignal = [];
    const signaledProcesses = [];
    const [executable, processArgs] = IS_WIN
        ? [
            "powershell.exe",
            [
                "-NoLogo",
                "-NoProfile",
                "-NonInteractive",
                "-Command",
                `& {
            [System.Console]::OutputEncoding = [System.Text.UTF8Encoding]::new($False, $False)
            Get-CimInstance -Class Win32_Process | ForEach-Object {
              "$($_.ParentProcessId) $($_.ProcessId) $($_.ExecutablePath)"
            }
          }`,
            ],
        ]
        : ["ps", ["-Ao", "ppid,pid,comm"]];
    const processOptions = {
        timeout: 5000,
        encoding: "utf-8",
    };
    for await (const line of invokeStdoutCliTool(executable, processArgs, processOptions)) {
        const match = PS_OUTPUT_PATTERN.exec(line);
        if (match) {
            const [, ppidStr, pidStr, comm] = match;
            const ppid = parseInt(ppidStr, 10);
            const entryPid = parseInt(pidStr, 10);
            const processBrief = {
                pid: entryPid,
                parentPid: ppid,
                command: comm,
            };
            if (entryPid === pid) {
                processesToSignal.push(processBrief);
            }
            const children = tree.get(ppid) ?? [];
            if (!children.length) {
                tree.set(ppid, children);
            }
            children.push(processBrief);
        }
    }
    const pidStack = [];
    for (let currentPid = pid; currentPid; currentPid = pidStack.shift()) {
        const children = tree.get(currentPid);
        if (children) {
            processesToSignal.push(...children);
            pidStack.push(...children.map(({ pid: childPid }) => childPid));
        }
    }
    for (const target of processesToSignal) {
        try {
            process.kill(target.pid, signal);
            signaledProcesses.push(target);
        }
        catch {
        }
    }
    return signaledProcesses;
};
