// @allurereport/web-commons/dist/stores/theme/store.js
import { computed, signal } from "@preact/signals-core";
import { getReportOptions } from "../../data.js";
import { persistSignal, restoreSignal } from "../persister/index.js";
import { STORAGE_KEY, THEME_AUTO, THEME_DARK, THEME_LIGHT } from "./constants.js";
import { getPrefersColorSchemeMQ, isAcceptedThemeValue, isAutoTheme } from "./utils.js";
const reportConfigTheme = getReportOptions()?.theme;
const getInitialPreferredTheme = () => {
    if (typeof window === "undefined") {
        return THEME_LIGHT;
    }
    if (getPrefersColorSchemeMQ().matches) {
        return THEME_DARK;
    }
    return THEME_LIGHT;
};
export const currentTheme = computed(() => {
    if (isAutoTheme(userTheme.value)) {
        return preferredTheme.value;
    }
    return userTheme.value;
});
export const userTheme = signal(THEME_AUTO);
export const preferredTheme = signal(getInitialPreferredTheme());
restoreSignal({
    signal: userTheme,
    key: STORAGE_KEY,
    onRestore: (value) => {
        if (isAcceptedThemeValue(value)) {
            return value;
        }
        return reportConfigTheme ?? THEME_AUTO;
    },
});
persistSignal({
    signal: userTheme,
    key: STORAGE_KEY,
});
export const themeStore = computed(() => ({
    current: currentTheme.value,
    selected: userTheme.value,
}));
