// @allurereport/plugin-testops/dist/logger.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Logger_instances, _Logger_level, _Logger_isLogLevel, _Logger_prefix;
import console from "node:console";
import { env } from "node:process";
import { inspect } from "node:util";
import ProgressBar from "progress";
import { blue, bold, cyan, dim, gray, red, yellow } from "yoctocolors";
const logLevelsPriority = {
    silent: Number.MAX_SAFE_INTEGER,
    verbose: 0,
    debug: 1,
    info: 2,
    warn: 3,
    error: 4,
};
function getLogLevelFromEnv() {
    if (env.LOG_LEVEL) {
        return env.LOG_LEVEL;
    }
    if (env.ALLURE_LOG_LEVEL) {
        return env.ALLURE_LOG_LEVEL;
    }
    if (env.NODE_ENV === "development") {
        return "debug";
    }
    return "info";
}
export class Logger {
    constructor(name) {
        _Logger_instances.add(this);
        this.name = name;
        _Logger_level.set(this, getLogLevelFromEnv());
        _Logger_prefix.set(this, cyan(bold(`[${this.name}]:`)));
    }
    progressBarCounter(message, total) {
        if (__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "silent")) {
            return {
                tick: () => { },
                update: () => { },
                terminate: () => { },
            };
        }
        const progressBar = new ProgressBar(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${blue(message)} [:bar] :current/:total`, {
            total,
            width: 20,
        });
        return progressBar;
    }
    progressBar(message) {
        if (__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "silent")) {
            return {
                update: () => { },
                terminate: () => { },
            };
        }
        const progressBar = new ProgressBar(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${blue(message)} [:bar] :percent`, {
            total: 100,
            width: 20,
        });
        return progressBar;
    }
    verbose(message) {
        if (!__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "verbose")) {
            return;
        }
        if (typeof message === "string") {
            console.log(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${dim(gray(message))}`);
            return;
        }
        console.log(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${dim(gray(JSON.stringify(message, null, 2)))}`);
    }
    debug(message) {
        if (!__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "debug")) {
            return;
        }
        if (typeof message === "string") {
            console.debug(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${gray(message)}`);
            return;
        }
        console.debug(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${gray(JSON.stringify(message, null, 2))}`);
    }
    newLine() {
        console.log("\n");
    }
    info(message) {
        if (!__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "info")) {
            return;
        }
        if (typeof message === "string") {
            console.info(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${blue(message)}`);
            return;
        }
        console.info(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${blue(JSON.stringify(message, null, 2))}`);
    }
    warn(message) {
        if (!__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "warn")) {
            return;
        }
        if (typeof message === "string") {
            console.warn(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${yellow(message)}`);
            return;
        }
        console.warn(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${yellow(JSON.stringify(message, null, 2))}`);
    }
    error(message) {
        if (!__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "error")) {
            return;
        }
        if (typeof message === "string") {
            console.error(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${red(message)}`);
            return;
        }
        console.error(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} ${red(JSON.stringify(message, null, 2))}`);
    }
    inspect(value) {
        if (!__classPrivateFieldGet(this, _Logger_instances, "m", _Logger_isLogLevel).call(this, "debug")) {
            return;
        }
        console.log(`${__classPrivateFieldGet(this, _Logger_prefix, "f")} inspect`);
        inspect(value);
    }
}
_Logger_level = new WeakMap(), _Logger_prefix = new WeakMap(), _Logger_instances = new WeakSet(), _Logger_isLogLevel = function _Logger_isLogLevel(level) {
    return logLevelsPriority[__classPrivateFieldGet(this, _Logger_level, "f")] <= logLevelsPriority[level];
};
