// @allurereport/plugin-testops/dist/plugin.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _TestOpsPlugin_instances, _TestOpsPlugin_logger, _TestOpsPlugin_ci, _TestOpsPlugin_client, _TestOpsPlugin_enabled, _TestOpsPlugin_launchName, _TestOpsPlugin_launchTags, _TestOpsPlugin_uploadedTestResultsIds, _TestOpsPlugin_autocloseLaunch, _TestOpsPlugin_uploadQualityGateResults, _TestOpsPlugin_uploadGlobalErrors, _TestOpsPlugin_uploadGlobalAttachments, _TestOpsPlugin_uploadTestResults, _TestOpsPlugin_upload, _TestOpsPlugin_trsToUpload, _TestOpsPlugin_enrichWithCategories, _TestOpsPlugin_syncLaunchCategories, _TestOpsPlugin_collectCategoryNamesByExternalId, _TestOpsPlugin_assignCreatedCategoryIds, _TestOpsPlugin_startUpload, _TestOpsPlugin_stopUpload;
import { env } from "node:process";
import { detect } from "@allurereport/ci";
import { getWorstStatus } from "@allurereport/core-api";
import { createPluginSummary } from "@allurereport/plugin-api";
import { uniqBy, stubTrue } from "lodash-es";
import { bold } from "yoctocolors";
import { TestOpsClient } from "./client.js";
import { Logger } from "./logger.js";
import { toUploadCategory } from "./uploadCategory.js";
import { uploadFilenameForLink } from "./uploaderDto.js";
import { attachmentsResolverFactory, fixturesResolverFactory, resolvePluginOptions, unwrapStepsAttachments, } from "./utils.js";
const categoryDisplayName = (cat) => cat.name ?? cat.grouping?.[0]?.name ?? cat.grouping?.[0]?.value ?? cat.grouping?.[0]?.key ?? cat.externalId;
export class TestOpsPlugin {
    constructor(options) {
        _TestOpsPlugin_instances.add(this);
        this.options = options;
        _TestOpsPlugin_logger.set(this, new Logger("TestOpsPlugin"));
        _TestOpsPlugin_ci.set(this, void 0);
        _TestOpsPlugin_client.set(this, void 0);
        _TestOpsPlugin_enabled.set(this, false);
        _TestOpsPlugin_launchName.set(this, "");
        _TestOpsPlugin_launchTags.set(this, []);
        _TestOpsPlugin_uploadedTestResultsIds.set(this, new Set());
        _TestOpsPlugin_autocloseLaunch.set(this, void 0);
        const { accessToken, endpoint, projectId, launchName, launchTags, autocloseLaunch = true, } = resolvePluginOptions(options);
        __classPrivateFieldSet(this, _TestOpsPlugin_ci, detect(), "f");
        if ([accessToken, endpoint, projectId].every(Boolean)) {
            __classPrivateFieldSet(this, _TestOpsPlugin_enabled, true, "f");
            __classPrivateFieldSet(this, _TestOpsPlugin_client, new TestOpsClient({
                baseUrl: endpoint,
                accessToken,
                projectId,
            }), "f");
            __classPrivateFieldSet(this, _TestOpsPlugin_launchName, launchName, "f");
            __classPrivateFieldSet(this, _TestOpsPlugin_launchTags, launchTags, "f");
        }
        __classPrivateFieldSet(this, _TestOpsPlugin_autocloseLaunch, autocloseLaunch, "f");
        if (!accessToken) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").warn(`Allure TestOps ${bold("access token")} is missing. Please provide a valid access token in the plugin options.`);
        }
        if (!endpoint) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").warn(`Allure TestOps ${bold("endpoint")} is missing. Please provide a valid endpoint in the plugin options.`);
        }
        if (!projectId) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").warn(`Allure TestOps ${bold("project ID")} is missing. Please provide a valid project ID in the plugin options.`);
        }
    }
    get ciMode() {
        return __classPrivateFieldGet(this, _TestOpsPlugin_ci, "f") && __classPrivateFieldGet(this, _TestOpsPlugin_ci, "f").type !== "local";
    }
    async start(context, store) {
        if (!__classPrivateFieldGet(this, _TestOpsPlugin_enabled, "f")) {
            return;
        }
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").verbose("Starting upload…");
        await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_startUpload).call(this);
        await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_upload).call(this, store, { issueNewToken: false, context, stage: "start" });
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info(`Allure TestOps Launch: ${__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchUrl}`);
    }
    async update(context, store) {
        if (!__classPrivateFieldGet(this, _TestOpsPlugin_enabled, "f")) {
            return;
        }
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").verbose("Updating (uploading new results)…");
        await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_upload).call(this, store, { context, stage: "update" });
    }
    async done(context, store) {
        if (!__classPrivateFieldGet(this, _TestOpsPlugin_enabled, "f")) {
            return;
        }
        const allTrs = await store.allTestResults({
            filter: this.options.filter,
            includeHidden: false,
        });
        const worstStatus = getWorstStatus(allTrs.map(({ status }) => status));
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").verbose("Finalizing upload…");
        await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_upload).call(this, store, { context, stage: "done" });
        await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_stopUpload).call(this, worstStatus || "unknown");
        const launchId = __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchId;
        if (typeof launchId !== "number") {
            return;
        }
        if (!__classPrivateFieldGet(this, _TestOpsPlugin_autocloseLaunch, "f")) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info(`Upload finished. Allure TestOps Launch: ${__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchUrl}`);
            return;
        }
        try {
            await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").closeLaunch(launchId);
        }
        catch (err) {
            if (err instanceof Error) {
                __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").debug(`Failed to close launch: ${err.message}`);
            }
            else {
                __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").debug("Failed to close launch");
            }
        }
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info(`Upload finished. Allure TestOps Launch: ${__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchUrl}`);
    }
    async info(context, store) {
        if (!__classPrivateFieldGet(this, _TestOpsPlugin_enabled, "f")) {
            return undefined;
        }
        if (!__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchUrl) {
            return undefined;
        }
        const summary = await createPluginSummary({
            name: __classPrivateFieldGet(this, _TestOpsPlugin_launchName, "f"),
            plugin: "TestOps",
            meta: {
                reportId: context.reportUuid,
            },
            filter: this.options.filter,
            history: context.history,
            ci: context.ci,
            store,
        });
        summary.remoteHref = __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchUrl;
        return summary;
    }
}
_TestOpsPlugin_logger = new WeakMap(), _TestOpsPlugin_ci = new WeakMap(), _TestOpsPlugin_client = new WeakMap(), _TestOpsPlugin_enabled = new WeakMap(), _TestOpsPlugin_launchName = new WeakMap(), _TestOpsPlugin_launchTags = new WeakMap(), _TestOpsPlugin_uploadedTestResultsIds = new WeakMap(), _TestOpsPlugin_autocloseLaunch = new WeakMap(), _TestOpsPlugin_instances = new WeakSet(), _TestOpsPlugin_uploadQualityGateResults = async function _TestOpsPlugin_uploadQualityGateResults(store) {
    const results = await store.qualityGateResults();
    const uniqueResults = uniqBy(results.filter(({ success }) => !success), ({ rule, environment }) => `${rule}-${environment}`);
    if (uniqueResults.length === 0) {
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").verbose("No quality gate results to upload");
        return;
    }
    const progressBar = __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").progressBar("Uploading quality gate results");
    try {
        progressBar.update(0);
        await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").uploadQualityGateResults(uniqueResults, (percent, total) => {
            progressBar.update(percent / total);
        });
        progressBar.update(1);
        progressBar.terminate();
    }
    catch (error) {
        progressBar.terminate();
        if (__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").isTestOpsClientError(error)) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error(`Failed to upload quality gate results: ${error.response.data.message}`);
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").debug(error.response?.data);
        }
        else if (error instanceof Error) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error(`Failed to upload quality gate results: ${error.message}`);
        }
        else {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error("Failed to upload quality gate results");
        }
    }
}, _TestOpsPlugin_uploadGlobalErrors = async function _TestOpsPlugin_uploadGlobalErrors(store) {
    const results = await store.allGlobalErrors();
    if (results.length === 0) {
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").verbose("No global errors to upload");
        return;
    }
    const progressBar = __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").progressBar("Uploading global errors");
    try {
        progressBar.update(0);
        await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").uploadGlobalErrors(results, (percent, total) => {
            progressBar.update(percent / total);
        });
        progressBar.update(1);
        progressBar.terminate();
    }
    catch (error) {
        progressBar.terminate();
        if (__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").isTestOpsClientError(error)) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error(`Failed to upload global errors: ${error.response.data.message}`);
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").debug(error.response?.data);
        }
        else if (error instanceof Error) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error(`Failed to upload global errors: ${error.message}`);
        }
        else {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error("Failed to upload global errors");
        }
    }
}, _TestOpsPlugin_uploadGlobalAttachments = async function _TestOpsPlugin_uploadGlobalAttachments(store) {
    const attachments = await store.allGlobalAttachments();
    if (attachments.length === 0) {
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").debug("No global attachments to upload");
        return;
    }
    const progressBar = __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").progressBar("Uploading global attachments");
    try {
        progressBar.update(0);
        await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").uploadGlobalAttachments({
            attachments,
            attachmentsResolver: async (attachmentLink) => {
                const content = await store.attachmentContentById(attachmentLink.id);
                const body = await content?.readContent(async (stream) => stream);
                const filename = uploadFilenameForLink(attachmentLink);
                if (filename === undefined || body === undefined) {
                    return undefined;
                }
                return {
                    originalFileName: filename,
                    contentType: attachmentLink.contentType ?? "application/octet-stream",
                    content: body,
                };
            },
            onProgress: (percent, total) => {
                progressBar.update(percent / total);
            },
        });
        progressBar.update(1);
        progressBar.terminate();
    }
    catch (error) {
        progressBar.terminate();
        if (__classPrivateFieldGet(this, _TestOpsPlugin_client, "f").isTestOpsClientError(error)) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error(`Failed to upload global attachments: ${error.response.data.message}`);
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").debug(error.response?.data);
        }
        else if (error instanceof Error) {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error(`Failed to upload global attachments: ${error.message}`);
        }
        else {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").error("Failed to upload global attachments");
        }
    }
}, _TestOpsPlugin_uploadTestResults = async function _TestOpsPlugin_uploadTestResults(store, trsToUpload, environments) {
    const totalCount = trsToUpload.length;
    __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info(`Preparing to upload ${bold(totalCount.toString())} ${totalCount > 1 ? "test results" : "test result"}`);
    const trsProgressBar = __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").progressBarCounter("Uploading test results", totalCount);
    const uploadedTrs = await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").uploadTestResults({
        attachmentsResolver: attachmentsResolverFactory(store),
        fixturesResolver: fixturesResolverFactory(store),
        environments,
        trs: trsToUpload,
        onProgress: () => trsProgressBar.tick(),
    });
    uploadedTrs.forEach((tr) => {
        __classPrivateFieldGet(this, _TestOpsPlugin_uploadedTestResultsIds, "f").add(tr.id);
    });
    const uploadedCount = uploadedTrs.length;
    trsProgressBar.update(uploadedCount / totalCount);
    trsProgressBar.terminate();
    if (uploadedCount === 0) {
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").warn("No test results were uploaded");
        return;
    }
    __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info(`Uploaded ${uploadedCount} ${uploadedCount > 1 ? "test results" : "test result"}`);
}, _TestOpsPlugin_upload = async function _TestOpsPlugin_upload(store, options = {}) {
    const { issueNewToken = true, context, stage } = options;
    const trsToUpload = await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_trsToUpload).call(this, store);
    if (trsToUpload.length === 0) {
        if (stage == "update") {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info("No new test results to upload");
        }
        if (stage === "done") {
            __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").info("No test results to upload");
        }
        return;
    }
    if (issueNewToken) {
        __classPrivateFieldGet(this, _TestOpsPlugin_logger, "f").verbose("Issuing new OAuth token");
        await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").issueOauthToken();
    }
    await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").createSession(env);
    await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_uploadGlobalAttachments).call(this, store);
    await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_uploadGlobalErrors).call(this, store);
    await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_uploadQualityGateResults).call(this, store);
    const environments = await store.allEnvironmentIdentities();
    const contextCategories = context?.categories ?? [];
    const trsEnrichedWithCategories = await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_enrichWithCategories).call(this, store, trsToUpload, contextCategories);
    await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_syncLaunchCategories).call(this, trsEnrichedWithCategories, contextCategories);
    await __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_uploadTestResults).call(this, store, trsEnrichedWithCategories, environments);
}, _TestOpsPlugin_trsToUpload = async function _TestOpsPlugin_trsToUpload(store) {
    const filter = this.options.filter ?? stubTrue;
    const filteredTrs = await store.allTestResults({
        filter: (tr) => {
            const uploaded = __classPrivateFieldGet(this, _TestOpsPlugin_uploadedTestResultsIds, "f").has(tr.id);
            if (uploaded) {
                return false;
            }
            return filter(tr);
        },
        includeHidden: false,
    });
    return filteredTrs;
}, _TestOpsPlugin_enrichWithCategories = async function _TestOpsPlugin_enrichWithCategories(store, trs, contextCategories) {
    return Promise.all(trs.map(async (tr) => {
        const environmentId = await store.environmentIdByTrId(tr.id);
        const base = {
            ...tr,
            ...(environmentId ? { environment: environmentId } : {}),
            steps: unwrapStepsAttachments(tr.steps),
        };
        const category = toUploadCategory(base, contextCategories ?? []);
        if (category) {
            base.category = category;
        }
        return base;
    }));
}, _TestOpsPlugin_syncLaunchCategories = async function _TestOpsPlugin_syncLaunchCategories(trs, contextCategories) {
    const categoryNamesByExternalId = __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_collectCategoryNamesByExternalId).call(this, trs);
    if (categoryNamesByExternalId.size === 0) {
        return;
    }
    const bulkItems = [];
    const seenExternalIds = new Set();
    for (const tr of trs) {
        const cat = tr.category;
        if (!cat?.externalId)
            continue;
        if (seenExternalIds.has(cat.externalId))
            continue;
        seenExternalIds.add(cat.externalId);
        bulkItems.push({
            externalId: cat.externalId,
            name: categoryNamesByExternalId.get(cat.externalId) ?? categoryDisplayName(cat),
            hide: cat.hide,
            expand: cat.expand,
        });
    }
    const rankByExternalId = new Map();
    for (const c of contextCategories) {
        if (!rankByExternalId.has(c.id)) {
            rankByExternalId.set(c.id, c.index);
        }
        if (!rankByExternalId.has(c.name)) {
            rankByExternalId.set(c.name, c.index);
        }
    }
    const ranked = bulkItems.map((item, i) => ({
        item,
        i,
        rank: rankByExternalId.get(item.externalId),
    }));
    ranked.sort((a, b) => {
        const ar = a.rank ?? Number.POSITIVE_INFINITY;
        const br = b.rank ?? Number.POSITIVE_INFINITY;
        if (ar !== br)
            return ar - br;
        return a.i - b.i;
    });
    const orderedBulkItems = ranked.map((r) => r.item);
    const launchId = __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").launchId;
    try {
        const created = await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").createLaunchCategoriesBulk(launchId, orderedBulkItems);
        const categoryIdByExternalId = new Map(created.map((r) => [r.externalId, r.id]));
        __classPrivateFieldGet(this, _TestOpsPlugin_instances, "m", _TestOpsPlugin_assignCreatedCategoryIds).call(this, trs, categoryIdByExternalId);
    }
    catch {
    }
}, _TestOpsPlugin_collectCategoryNamesByExternalId = function _TestOpsPlugin_collectCategoryNamesByExternalId(trs) {
    const map = new Map();
    for (const tr of trs) {
        const cat = tr.category;
        if (cat?.externalId) {
            map.set(cat.externalId, categoryDisplayName(cat));
        }
    }
    return map;
}, _TestOpsPlugin_assignCreatedCategoryIds = function _TestOpsPlugin_assignCreatedCategoryIds(trs, idByExternalId) {
    for (const tr of trs) {
        const cat = tr.category;
        if (!cat?.externalId) {
            continue;
        }
        const id = idByExternalId.get(cat.externalId);
        if (typeof id === "number") {
            tr.category = { ...cat, id };
        }
    }
}, _TestOpsPlugin_startUpload = async function _TestOpsPlugin_startUpload() {
    await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").issueOauthToken();
    await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").createLaunch(__classPrivateFieldGet(this, _TestOpsPlugin_launchName, "f"), __classPrivateFieldGet(this, _TestOpsPlugin_launchTags, "f"));
    if (!this.ciMode) {
        return;
    }
    await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").startUpload(__classPrivateFieldGet(this, _TestOpsPlugin_ci, "f"));
}, _TestOpsPlugin_stopUpload = async function _TestOpsPlugin_stopUpload(status) {
    if (!this.ciMode) {
        return;
    }
    await __classPrivateFieldGet(this, _TestOpsPlugin_client, "f").stopUpload(__classPrivateFieldGet(this, _TestOpsPlugin_ci, "f"), status);
};
