// @allurereport/plugin-testops/dist/utils.js
import { env } from "node:process";
import { uploadFilenameForLink } from "./uploaderDto.js";
export const unwrapStepsAttachments = (steps) => {
    return steps.map((step) => {
        if (step.type === "attachment") {
            return {
                ...step,
                attachment: step.link,
            };
        }
        if (step.steps) {
            return {
                ...step,
                steps: unwrapStepsAttachments(step.steps),
            };
        }
        return step;
    });
};
export const resolvePluginOptions = (options) => {
    const { ALLURE_TOKEN, ALLURE_ENDPOINT, ALLURE_PROJECT_ID, ALLURE_LAUNCH_TAGS, ALLURE_LAUNCH_NAME } = env;
    const { accessToken = ALLURE_TOKEN, endpoint = ALLURE_ENDPOINT, projectId = ALLURE_PROJECT_ID, launchTags = ALLURE_LAUNCH_TAGS, launchName = ALLURE_LAUNCH_NAME, autocloseLaunch, } = options;
    const tags = !launchTags
        ? []
        : Array.isArray(launchTags)
            ? launchTags
            : launchTags.split(",").map((tag) => tag.trim());
    return {
        launchName: launchName || "Allure Report",
        launchTags: tags,
        accessToken: accessToken || "",
        endpoint: endpoint || "",
        projectId: projectId || "",
        ...(autocloseLaunch !== undefined ? { autocloseLaunch } : {}),
    };
};
export function attachmentsResolverFactory(store) {
    const attachmentsResolver = async (tr) => {
        const attachments = await store.attachmentsByTrId(tr.id);
        const result = [];
        await Promise.all(attachments.map(async (attachment) => {
            const content = await store.attachmentContentById(attachment.id);
            const body = await content?.readContent(async (s) => s);
            const filename = uploadFilenameForLink(attachment);
            if (filename === undefined || body === undefined) {
                return undefined;
            }
            result.push({
                originalFileName: filename,
                contentType: attachment.contentType ?? "application/octet-stream",
                content: body,
            });
        }));
        return result;
    };
    return attachmentsResolver;
}
export function fixturesResolverFactory(store) {
    const fixturesResolver = async (tr) => {
        const fixtures = await store.fixturesByTrId(tr.id);
        return fixtures.map((fxt) => ({
            ...fxt,
            type: fxt.type.toUpperCase(),
            steps: unwrapStepsAttachments(fxt.steps),
        }));
    };
    return fixturesResolver;
}
