// @allurereport/charts-api/dist/types.js
export var ChartType;
(function (ChartType) {
    ChartType["CurrentStatus"] = "currentStatus";
    ChartType["StatusDynamics"] = "statusDynamics";
    ChartType["StatusTransitions"] = "statusTransitions";
    ChartType["StabilityDistribution"] = "stabilityDistribution";
    ChartType["TestBaseGrowthDynamics"] = "testBaseGrowthDynamics";
    ChartType["StatusAgePyramid"] = "statusAgePyramid";
    ChartType["Durations"] = "durations";
    ChartType["DurationDynamics"] = "durationDynamics";
    ChartType["TrSeverities"] = "testResultSeverities";
    ChartType["TestingPyramid"] = "testingPyramid";
    ChartType["CoverageDiff"] = "coverageDiff";
    ChartType["SuccessRateDistribution"] = "successRateDistribution";
    ChartType["ProblemsDistribution"] = "problemsDistribution";
})(ChartType || (ChartType = {}));
export var ChartDataType;
(function (ChartDataType) {
    ChartDataType["Status"] = "status";
    ChartDataType["Severity"] = "severity";
})(ChartDataType || (ChartDataType = {}));
export var ChartMode;
(function (ChartMode) {
    ChartMode["Raw"] = "raw";
    ChartMode["Percent"] = "percent";
    ChartMode["Diverging"] = "diverging";
})(ChartMode || (ChartMode = {}));
