// @allurereport/web-commons/dist/stores/url/store.js
import { computed, signal } from "@preact/signals-core";
import { getCurrentUrl, goTo, searchParamsToParams, subscribeToUrlChange } from "./helpers.js";
export const currentUrlSignal = signal(getCurrentUrl());
subscribeToUrlChange(() => {
    if (currentUrlSignal.peek() === getCurrentUrl()) {
        return;
    }
    currentUrlSignal.value = getCurrentUrl();
});
const urlSignal = computed(() => new URL(currentUrlSignal.value));
export const urlSearchParams = computed(() => urlSignal.value.searchParams);
export const setParams = (...params) => {
    const newUrl = new URL(getCurrentUrl());
    for (const param of params) {
        newUrl.searchParams.delete(param.key);
        if (Array.isArray(param.value)) {
            for (const value of param.value) {
                newUrl.searchParams.append(param.key, value);
            }
        }
        else if (typeof param.value === "string") {
            newUrl.searchParams.set(param.key, param.value);
        }
    }
    if (newUrl.href === urlSignal.peek().href) {
        return;
    }
    goTo(newUrl.href, {
        replace: true,
    });
};
export const currentUrl = computed(() => {
    return {
        hash: urlSignal.value.hash,
        pathname: urlSignal.value.pathname,
        origin: urlSignal.value.origin,
        params: searchParamsToParams(urlSearchParams.value),
    };
});
export const getParamValue = (key) => urlSearchParams.value.get(key);
export const getParamValues = (key) => urlSearchParams.value.getAll(key);
export const hasParam = (key) => urlSearchParams.value.has(key);
export const getParamValueComputed = (key) => computed(() => getParamValue(key));
export const getParamValuesComputed = (key) => computed(() => getParamValues(key));
export const hasParamComputed = (key) => computed(() => hasParam(key));
