// @allurereport/plugin-jira/dist/helpers.js
import { DEFAULT_ENVIRONMENT } from "@allurereport/core-api";
const isJiraUrl = (url) => {
    const jiraPattern = /^https:\/\/[a-zA-Z0-9-]+\.atlassian\.net\/browse\/[A-Z]+-\d+$/;
    return jiraPattern.test(url);
};
export const isJiraIssueKey = (issue) => {
    const jiraPattern = /^[A-Z]+-\d+$/;
    return jiraPattern.test(issue);
};
export const findJiraLink = (tr) => {
    return tr.links.find((link) => isJiraUrl(link.url));
};
const filterKeyParams = (params, runsCount) => {
    const result = [];
    const intermediate = new Map();
    for (const p of params) {
        if (p.excluded || p.hidden) {
            continue;
        }
        if (!intermediate.has(p.name)) {
            intermediate.set(p.name, { value: p.value, count: 1 });
            continue;
        }
        if (intermediate.get(p.name).value !== p.value) {
            intermediate.delete(p.name);
            continue;
        }
        intermediate.get(p.name).count++;
    }
    for (const [name, value] of intermediate) {
        if (value.count === runsCount) {
            result.push({ name, value: value.value });
            continue;
        }
    }
    return result;
};
const filterOutDefaultEnvironment = (env) => {
    if (env === DEFAULT_ENVIRONMENT) {
        return undefined;
    }
    return env;
};
export const prepareTestResults = (trs) => {
    const trMap = new Map();
    for (const tr of trs) {
        const jiraLink = findJiraLink(tr);
        if (!jiraLink) {
            continue;
        }
        const trId = tr.historyId ?? tr.id;
        if (!trMap.has(trId)) {
            trMap.set(trId, {
                id: trId,
                entries: [],
                issue: jiraLink,
                name: trimName(tr.name),
                keyParams: [],
            });
        }
        const storedTr = trMap.get(trId);
        storedTr.entries.push({ status: tr.status, env: filterOutDefaultEnvironment(tr.environment), date: tr.stop });
        storedTr.keyParams.push(...tr.parameters);
    }
    return Array.from(trMap.values()).map((tr) => {
        return {
            ...tr,
            keyParams: filterKeyParams(tr.keyParams, tr.entries.length),
        };
    });
};
const trimStrMax = (str, maxLength = 255) => {
    if (str.length <= maxLength) {
        return str;
    }
    const trimmed = str.slice(0, maxLength);
    return `${trimmed.replace(/\.+$/, "")}...`;
};
export const trimName = (name) => trimStrMax(name, 255);
export const trimParameters = (p) => trimStrMax(p, 120);
export const trimCiInfoLabel = (label) => trimStrMax(label, 120);
