// @allurereport/plugin-jira/dist/plugin.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _JiraPlugin_instances, _JiraPlugin_pluginName, _JiraPlugin_pluginOptions_get, _JiraPlugin_verifyOptions, _JiraPlugin_requestForgeApp, _JiraPlugin_uploadResults, _JiraPlugin_getReportStatus, _JiraPlugin_getStatisticByEnv, _JiraPlugin_getCiInfo, _JiraPlugin_getReportDate, _JiraPlugin_uploadReport, _JiraPlugin_uploadAll;
import { DEFAULT_ENVIRONMENT, getWorstStatus, statusesList } from "@allurereport/core-api";
import axios, { isAxiosError } from "axios";
import { isJiraIssueKey, prepareTestResults, trimCiInfoLabel, trimName } from "./helpers.js";
const TRUE_VALUES = ["true", "1"];
export class JiraPlugin {
    constructor(options = {}) {
        _JiraPlugin_instances.add(this);
        this.options = options;
        _JiraPlugin_pluginName.set(this, "Allure Jira Plugin");
    }
    async clearReports(issues) {
        const payload = { issues, reports: true };
        return await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_requestForgeApp).call(this, { operation: "clear", payload });
    }
    async clearResults(issues) {
        return await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_requestForgeApp).call(this, { operation: "clear", payload: { issues, results: true } });
    }
    async clearAll(issues) {
        const payload = { issues, reports: true, results: true };
        return await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_requestForgeApp).call(this, { operation: "clear", payload });
    }
    async done(context, store) {
        __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_verifyOptions).call(this);
        const { uploadReport, uploadResults } = __classPrivateFieldGet(this, _JiraPlugin_instances, "a", _JiraPlugin_pluginOptions_get);
        if (!uploadReport && !uploadResults) {
            throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] Set at least one of the options: uploadReport or uploadResults`);
        }
        if (uploadReport && uploadResults) {
            await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_uploadAll).call(this, context, store);
            return;
        }
        if (uploadReport) {
            await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_uploadReport).call(this, context, store);
            return;
        }
        if (uploadResults) {
            await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_uploadResults).call(this, context, store);
            return;
        }
    }
}
_JiraPlugin_pluginName = new WeakMap(), _JiraPlugin_instances = new WeakSet(), _JiraPlugin_pluginOptions_get = function _JiraPlugin_pluginOptions_get() {
    return {
        token: this.options.token || process.env.ALLURE_JIRA_TOKEN,
        webhook: this.options.webhook || process.env.ALLURE_JIRA_WEBHOOK,
        reportIssue: this.options.issue ?? process.env.ALLURE_JIRA_ISSUE,
        uploadReport: this.options.uploadReport ?? TRUE_VALUES.includes(process.env.ALLURE_JIRA_UPLOAD_REPORT ?? ""),
        uploadResults: this.options.uploadResults ?? TRUE_VALUES.includes(process.env.ALLURE_JIRA_UPLOAD_RESULTS ?? ""),
    };
}, _JiraPlugin_verifyOptions = function _JiraPlugin_verifyOptions() {
    const { token, webhook } = __classPrivateFieldGet(this, _JiraPlugin_instances, "a", _JiraPlugin_pluginOptions_get);
    if (!token) {
        throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] token is not set`);
    }
    if (!webhook) {
        throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] webhook is not set`);
    }
}, _JiraPlugin_requestForgeApp = async function _JiraPlugin_requestForgeApp(props) {
    const { operation, payload, version = "v1" } = props;
    const { token, webhook } = __classPrivateFieldGet(this, _JiraPlugin_instances, "a", _JiraPlugin_pluginOptions_get);
    const requestData = {
        operation,
        version,
        payload,
        token: token,
    };
    try {
        await axios.post(webhook, requestData);
    }
    catch (error) {
        if (isAxiosError(error)) {
            const errorMessage = error.response?.data?.message || error.message;
            throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] Allure Jira Integration app error: ${errorMessage}`);
        }
        throw error;
    }
}, _JiraPlugin_uploadResults = async function _JiraPlugin_uploadResults(context, store) {
    const allTestResults = await store.allTestResults();
    if (allTestResults.length === 0) {
        throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] no test results found`);
    }
    const testResults = prepareTestResults(allTestResults);
    const payload = {
        results: testResults,
        reportUrl: context.reportUrl,
    };
    await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_requestForgeApp).call(this, { operation: "upload-results", payload });
}, _JiraPlugin_getReportStatus = async function _JiraPlugin_getReportStatus(store, statistic) {
    const globalErrors = await store.allGlobalErrors();
    const globalExitCode = await store.globalExitCode();
    const code = globalExitCode?.actual ?? globalExitCode?.original;
    const hasGlobalErrors = globalErrors.length > 0;
    if (hasGlobalErrors) {
        return "failed";
    }
    if (code === 0) {
        return "passed";
    }
    const worstStatus = getWorstStatus(statusesList
        .map((status) => {
        const value = statistic[status] ?? 0;
        return value > 0 ? status : undefined;
    })
        .filter((status) => status !== undefined)) ?? "passed";
    if (worstStatus === "passed") {
        return "passed";
    }
    return "failed";
}, _JiraPlugin_getStatisticByEnv = async function _JiraPlugin_getStatisticByEnv(store) {
    const statisticByEnv = {};
    const envs = await store.allEnvironments();
    for (const env of envs) {
        if (env === DEFAULT_ENVIRONMENT) {
            continue;
        }
        statisticByEnv[env] = await store.testsStatistic((tr) => tr.environment === env);
    }
    return statisticByEnv;
}, _JiraPlugin_getCiInfo = async function _JiraPlugin_getCiInfo(context) {
    const jobUrl = context.ci?.pullRequestUrl ?? context.ci?.jobUrl ?? context.ci?.jobRunUrl;
    const jobLabel = context.ci?.pullRequestName ?? context.ci?.jobName ?? context.ci?.jobRunName;
    return { url: jobUrl, label: jobLabel ? trimCiInfoLabel(jobLabel) : undefined };
}, _JiraPlugin_getReportDate = async function _JiraPlugin_getReportDate(store) {
    const trs = await store.allTestResults();
    return trs.reduce((acc, { stop }) => Math.max(acc, stop || 0), 0);
}, _JiraPlugin_uploadReport = async function _JiraPlugin_uploadReport(context, store) {
    const { reportIssue } = __classPrivateFieldGet(this, _JiraPlugin_instances, "a", _JiraPlugin_pluginOptions_get);
    if (!reportIssue) {
        throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] reportIssue is not set`);
    }
    if (!isJiraIssueKey(reportIssue)) {
        throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] reportIssue is not a valid Jira issue key`);
    }
    const statistic = await store.testsStatistic();
    if (statistic.total === 0) {
        throw new Error(`[${__classPrivateFieldGet(this, _JiraPlugin_pluginName, "f")}] no test results found`);
    }
    const history = await store.allHistoryDataPoints();
    const reportStatus = await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_getReportStatus).call(this, store, statistic);
    const statisticByEnv = await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_getStatisticByEnv).call(this, store);
    const date = await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_getReportDate).call(this, store);
    const ciInfo = await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_getCiInfo).call(this, context);
    const payload = {
        issue: reportIssue,
        report: {
            id: context.reportUuid,
            history: history.map(({ uuid }) => uuid),
            status: reportStatus,
            name: trimName(context.reportName),
            url: context.reportUrl,
            date,
            ciInfo: ciInfo.url ? { url: ciInfo.url, label: ciInfo.label } : undefined,
            statistic,
            statisticByEnv,
        },
    };
    await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_requestForgeApp).call(this, { operation: "upload-report", payload });
}, _JiraPlugin_uploadAll = async function _JiraPlugin_uploadAll(context, store) {
    await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_uploadReport).call(this, context, store);
    await __classPrivateFieldGet(this, _JiraPlugin_instances, "m", _JiraPlugin_uploadResults).call(this, context, store);
};
