// @allurereport/aql/dist/errors/index.js
/* eslint-disable max-len */
/**
 * AQL error types and classes
 */
export var AqlErrorCode;
(function (AqlErrorCode) {
    /** Tokenizer errors */
    AqlErrorCode["UNEXPECTED_CHARACTER"] = "UNEXPECTED_CHARACTER";
    AqlErrorCode["UNTERMINATED_STRING"] = "UNTERMINATED_STRING";
    AqlErrorCode["INVALID_UNICODE_ESCAPE"] = "INVALID_UNICODE_ESCAPE";
    /** Parser errors */
    AqlErrorCode["EXPECTED_TOKEN"] = "EXPECTED_TOKEN";
    AqlErrorCode["EXPECTED_OPERATION"] = "EXPECTED_OPERATION";
    AqlErrorCode["EXPECTED_VALUE"] = "EXPECTED_VALUE";
    AqlErrorCode["EXPECTED_ACCESSOR"] = "EXPECTED_ACCESSOR";
    AqlErrorCode["INVALID_SYNTAX"] = "INVALID_SYNTAX";
    AqlErrorCode["INVALID_INPUT"] = "INVALID_INPUT";
    AqlErrorCode["INVALID_IDENTIFIER"] = "INVALID_IDENTIFIER";
    /** Configuration errors */
    AqlErrorCode["FORBIDDEN_LOGICAL_OPERATOR"] = "FORBIDDEN_LOGICAL_OPERATOR";
    AqlErrorCode["FORBIDDEN_OPERATION"] = "FORBIDDEN_OPERATION";
    AqlErrorCode["FORBIDDEN_ARRAY_OPERATION"] = "FORBIDDEN_ARRAY_OPERATION";
    AqlErrorCode["FORBIDDEN_IDENTIFIER"] = "FORBIDDEN_IDENTIFIER";
    AqlErrorCode["FORBIDDEN_VALUE_TYPE"] = "FORBIDDEN_VALUE_TYPE";
    AqlErrorCode["FORBIDDEN_PARENTHESES"] = "FORBIDDEN_PARENTHESES";
    AqlErrorCode["FORBIDDEN_BRACKET_ACCESS"] = "FORBIDDEN_BRACKET_ACCESS";
})(AqlErrorCode || (AqlErrorCode = {}));
/**
 * Base error class for AQL errors
 */
export class AqlError extends Error {
    constructor(code, message, details = {}) {
        super(message);
        this.name = "AqlError";
        this.code = code;
        this.details = details;
        /**
         * Maintains proper stack trace for where our error was thrown (only available on V8)
         */
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, this.constructor);
        }
    }
    /**
     * Gets error details for translation/localization
     * Includes error code and all context information
     */
    get fullDetails() {
        return { ...this.details, code: this.code };
    }
    /**
     * Gets error code for translation key generation
     */
    get translationKey() {
        return `aql.errors.${this.code}`;
    }
}
/**
 * Tokenizer error
 */
export class AqlTokenizerError extends AqlError {
    constructor(code, message, details = {}) {
        super(code, message, details);
        this.name = "AqlTokenizerError";
    }
}
/**
 * Parser error
 */
export class AqlParserError extends AqlError {
    constructor(code, message, details = {}) {
        super(code, message, details);
        this.name = "AqlParserError";
    }
}
/**
 * Helper functions to create specific errors
 */
export const AqlErrors = {
    unexpectedCharacter: (character, position) => {
        return new AqlTokenizerError(AqlErrorCode.UNEXPECTED_CHARACTER, `Unexpected character: ${character} at position ${position}`, { character, position });
    },
    unterminatedString: (position) => {
        return new AqlTokenizerError(AqlErrorCode.UNTERMINATED_STRING, `Unterminated string at position ${position}`, {
            position,
        });
    },
    invalidUnicodeEscape: (position) => {
        return new AqlTokenizerError(AqlErrorCode.INVALID_UNICODE_ESCAPE, `Invalid Unicode escape sequence at position ${position}`, { position });
    },
    expectedToken: (expected, got, position, context) => {
        const message = context
            ? `Expected ${expected}, got ${got} at position ${position}. Context: ${context}`
            : `Expected ${expected}, got ${got} at position ${position}`;
        return new AqlParserError(AqlErrorCode.EXPECTED_TOKEN, message, {
            expected,
            got,
            position,
            context,
        });
    },
    expectedOperation: (position) => {
        return new AqlParserError(AqlErrorCode.EXPECTED_OPERATION, `Expected operation at position ${position}`, {
            position,
        });
    },
    expectedValue: (position) => {
        return new AqlParserError(AqlErrorCode.EXPECTED_VALUE, `Expected value at position ${position}`, { position });
    },
    expectedAccessor: (position, expected) => {
        const message = expected
            ? `Expected ${expected} in accessor at position ${position}`
            : `Expected STRING or NUMBER in accessor at position ${position}`;
        return new AqlParserError(AqlErrorCode.EXPECTED_ACCESSOR, message, {
            position,
            expected,
        });
    },
    invalidInput: (reason) => {
        return new AqlParserError(AqlErrorCode.INVALID_INPUT, reason, { reason });
    },
    invalidIdentifier: (identifier, position) => {
        return new AqlParserError(AqlErrorCode.INVALID_IDENTIFIER, `Invalid identifier '${identifier}' at position ${position}. Identifier must contain only Latin letters (a-z, A-Z) and underscores (_)`, { identifier, position });
    },
    forbiddenLogicalOperator: (operator, position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_LOGICAL_OPERATOR, `Logical operator '${operator}' is not allowed at position ${position}`, { operator, position });
    },
    forbiddenOperation: (operation, position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_OPERATION, `Operation '${operation}' is not allowed at position ${position}`, { operation, position });
    },
    forbiddenArrayOperation: (operation, position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_ARRAY_OPERATION, `Array operation '${operation}' is not allowed at position ${position}`, { operation, position });
    },
    forbiddenIdentifier: (identifier, position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_IDENTIFIER, `Identifier '${identifier}' is not allowed at position ${position}`, { identifier, position });
    },
    forbiddenValueType: (valueType, position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_VALUE_TYPE, `Value type '${valueType}' is not allowed at position ${position}`, { valueType, position });
    },
    forbiddenParentheses: (position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_PARENTHESES, `Parentheses are not allowed at position ${position}`, { position });
    },
    forbiddenBracketAccess: (position) => {
        return new AqlParserError(AqlErrorCode.FORBIDDEN_BRACKET_ACCESS, `Bracket access is not allowed at position ${position}`, { position });
    },
};
/**
 * Type guard to check if error is AqlError
 */
export function isAqlError(error) {
    return error instanceof AqlError;
}
/**
 * Type guard to check if error is AqlTokenizerError
 */
export function isAqlTokenizerError(error) {
    return error instanceof AqlTokenizerError;
}
/**
 * Type guard to check if error is AqlParserError
 */
export function isAqlParserError(error) {
    return error instanceof AqlParserError;
}
