// @allurereport/ci/dist/detectors/azure.js
import { CiType } from "@allurereport/core-api";
import { getEnv } from "../utils.js";
export const getRootURL = () => getEnv("SYSTEM_COLLECTIONURI");
export const getBuildID = () => getEnv("BUILD_BUILDID");
export const getDefinitionID = () => getEnv("SYSTEM_DEFINITIONID");
export const getProjectID = () => getEnv("SYSTEM_TEAMPROJECTID");
export const azure = {
    type: CiType.Azure,
    get detected() {
        return getEnv("SYSTEM_DEFINITIONID") !== "";
    },
    get repoName() {
        const repoName = getEnv("BUILD_REPOSITORY_NAME");
        return repoName.split("/")?.[1] ?? repoName;
    },
    get jobUid() {
        return `${getProjectID()}_${getDefinitionID()}`;
    },
    get jobUrl() {
        return `${getRootURL()}/${getProjectID()}/_build?definitionId=${getDefinitionID()}`;
    },
    get jobName() {
        return getEnv("BUILD_DEFINITIONNAME");
    },
    get jobRunUid() {
        return getBuildID();
    },
    get jobRunUrl() {
        return `${getRootURL()}/${getProjectID()}/_build/results?buildId=${getBuildID()}`;
    },
    get jobRunName() {
        return getEnv("BUILD_BUILDNUMBER");
    },
    get jobRunBranch() {
        return getEnv("BUILD_SOURCEBRANCHNAME");
    },
    get pullRequestUrl() {
        const repositoryProvider = getEnv("BUILD_REPOSITORY_PROVIDER");
        const repositoryUrl = getEnv("SYSTEM_PULLREQUEST_SOURCEREPOSITORYURI");
        const pullRequestNumber = getEnv("SYSTEM_PULLREQUEST_PULLREQUESTNUMBER");
        if (repositoryProvider === "GitHub") {
            return `${repositoryUrl}/pull/${pullRequestNumber}`;
        }
        if (repositoryProvider === "TfsGit" || repositoryProvider === "TfsVersionControl") {
            return `${repositoryUrl}/pullrequest/${pullRequestNumber}`;
        }
        return "";
    },
    get pullRequestName() {
        return "";
    },
};
