// @allurereport/ci/dist/detectors/bitbucket.js
import { CiType } from "@allurereport/core-api";
import { getEnv } from "../utils.js";
export const getJobURL = () => {
    const origin = getEnv("BITBUCKET_GIT_HTTP_ORIGIN");
    return `${origin}/pipelines`;
};
export const bitbucket = {
    type: CiType.Bitbucket,
    get detected() {
        return getEnv("BITBUCKET_PIPELINE_UUID") !== "";
    },
    get repoName() {
        return getEnv("BITBUCKET_REPO_SLUG");
    },
    get jobUid() {
        return getEnv("BITBUCKET_REPO_FULL_NAME");
    },
    get jobUrl() {
        return getJobURL();
    },
    get jobName() {
        return getEnv("BITBUCKET_REPO_FULL_NAME");
    },
    get jobRunUid() {
        return getEnv("BITBUCKET_PIPELINE_UUID");
    },
    get jobRunUrl() {
        return `${getJobURL()}/results/${this.jobRunUid}`;
    },
    get jobRunName() {
        return getEnv("BITBUCKET_PIPELINE_UUID");
    },
    get jobRunBranch() {
        return getEnv("BITBUCKET_BRANCH");
    },
    get pullRequestUrl() {
        const prId = getEnv("BITBUCKET_PR_ID");
        if (!prId) {
            return "";
        }
        const origin = getEnv("BITBUCKET_GIT_HTTP_ORIGIN");
        return `${origin}/pull-requests/${prId}`;
    },
    get pullRequestName() {
        return "";
    },
};
