// @allurereport/ci/dist/detectors/circle.js
import { CiType } from "@allurereport/core-api";
import { getEnv, parseURLPath } from "../utils.js";
export const getBuildNumber = () => getEnv("CIRCLE_BUILD_NUM");
export const getJobRunURL = () => getEnv("CIRCLE_BUILD_URL");
const getJobURL = () => {
    const jobRunURL = getJobRunURL();
    const buildNumber = getBuildNumber();
    return jobRunURL.replace(`/${buildNumber}`, "");
};
export const circle = {
    type: CiType.Circle,
    get detected() {
        const hasEnv = getEnv("CIRCLECI") !== "";
        const jobURL = getJobURL();
        const path = parseURLPath(jobURL);
        return hasEnv && path !== "";
    },
    get repoName() {
        return getEnv("CIRCLE_PROJECT_REPONAME");
    },
    get jobUid() {
        const jobURL = getJobURL();
        return parseURLPath(jobURL);
    },
    get jobUrl() {
        return getJobURL();
    },
    get jobName() {
        const username = getEnv("CIRCLE_USERNAME");
        const reponame = getEnv("CIRCLE_PROJECT_REPONAME");
        return `${username}/${reponame}`;
    },
    get jobRunUid() {
        return getEnv("CIRCLE_WORKFLOW_JOB_ID");
    },
    get jobRunUrl() {
        return getJobRunURL();
    },
    get jobRunName() {
        return getEnv("CIRCLE_BUILD_NUM");
    },
    get jobRunBranch() {
        return getEnv("CIRCLE_BRANCH");
    },
    get pullRequestUrl() {
        return "";
    },
    get pullRequestName() {
        return "";
    },
};
