// @allurereport/ci/dist/detectors/gitlab.js
import { CiType } from "@allurereport/core-api";
import { getEnv } from "../utils.js";
export const gitlab = {
    type: CiType.Gitlab,
    get detected() {
        return getEnv("GITLAB_CI") !== "";
    },
    get repoName() {
        return getEnv("CI_PROJECT_NAME");
    },
    get jobUid() {
        return getEnv("CI_PROJECT_ID");
    },
    get jobUrl() {
        return `${getEnv("CI_PROJECT_URL")}/pipelines`;
    },
    get jobName() {
        return getEnv("CI_PROJECT_NAME");
    },
    get jobRunUid() {
        return getEnv("CI_PIPELINE_ID");
    },
    get jobRunUrl() {
        return getEnv("CI_PIPELINE_URL");
    },
    get jobRunName() {
        return getEnv("CI_PIPELINE_ID");
    },
    get jobRunBranch() {
        return getEnv("CI_COMMIT_REF_NAME");
    },
    get pullRequestUrl() {
        const mergeRequestIID = getEnv("CI_MERGE_REQUEST_IID");
        if (!mergeRequestIID) {
            return "";
        }
        const projectUrl = getEnv("CI_PROJECT_URL");
        return `${projectUrl}/-/merge_requests/${mergeRequestIID}`;
    },
    get pullRequestName() {
        return getEnv("CI_MERGE_REQUEST_TITLE");
    },
};
