// @allurereport/ci/dist/detectors/jenkins.js
import { CiType } from "@allurereport/core-api";
import { getEnv, getReponameFromRepoUrl } from "../utils.js";
export const jenkins = {
    type: CiType.Jenkins,
    get detected() {
        return getEnv("JENKINS_URL") !== "";
    },
    get repoName() {
        const gitUrl = getEnv("GIT_URL");
        if (!gitUrl) {
            return "";
        }
        return getReponameFromRepoUrl(gitUrl);
    },
    get jobUid() {
        return getEnv("JOB_NAME");
    },
    get jobUrl() {
        return getEnv("JOB_URL");
    },
    get jobName() {
        return getEnv("JOB_BASE_NAME");
    },
    get jobRunUid() {
        return getEnv("BUILD_NUMBER");
    },
    get jobRunUrl() {
        return getEnv("BUILD_URL");
    },
    get jobRunName() {
        return getEnv("BUILD_DISPLAY_NAME");
    },
    get jobRunBranch() {
        return getEnv("BRANCH_NAME");
    },
    get pullRequestUrl() {
        return getEnv("CHANGE_URL");
    },
    get pullRequestName() {
        return getEnv("CHANGE_TITLE");
    },
};
