// @allurereport/ci/dist/detectors/local.js
import { spawnSync } from "node:child_process";
import { basename } from "node:path";
import { CiType } from "@allurereport/core-api";
export const local = {
    type: CiType.Local,
    get detected() {
        return true;
    },
    get repoName() {
        const output = spawnSync("git", ["rev-parse", "--show-toplevel"]);
        if (output.error) {
            return "";
        }
        return basename(output.stdout.toString().trim());
    },
    get jobUid() {
        return "";
    },
    get jobUrl() {
        return "";
    },
    get jobName() {
        return "";
    },
    get jobRunUid() {
        return "";
    },
    get jobRunUrl() {
        return "";
    },
    get jobRunName() {
        return "";
    },
    get jobRunBranch() {
        const output = spawnSync("git", ["rev-parse", "--abbrev-ref", "HEAD"]);
        if (output.error) {
            return "";
        }
        return output.stdout.toString().trim();
    },
    get pullRequestUrl() {
        return "";
    },
    get pullRequestName() {
        return "";
    },
};
