// @allurereport/core-api/dist/constants.js
export const statusesList = ["failed", "broken", "passed", "skipped", "unknown"];
export const severityLevels = ["blocker", "critical", "normal", "minor", "trivial"];
export const severityLabelName = "severity";
export const fallbackTestCaseIdLabelName = "_fallbackTestCaseId";
export const unsuccessfulStatuses = new Set(["failed", "broken"]);
export const successfulStatuses = new Set(["passed"]);
export const includedInSuccessRate = new Set([...unsuccessfulStatuses, ...successfulStatuses]);
export const filterByStatus = (statuses) => {
    const set = new Set(statuses);
    return (t) => set.has(t.status);
};
export const filterSuccessful = filterByStatus(successfulStatuses);
export const filterUnsuccessful = filterByStatus(unsuccessfulStatuses);
export const filterIncludedInSuccessRate = filterByStatus(includedInSuccessRate);
export const emptyStatistic = () => ({ total: 0 });
export const incrementStatistic = (statistic, status, count = 1) => {
    statistic[status] = (statistic?.[status] ?? 0) + count;
    statistic.total += count;
};
export const mergeStatistic = (statistic, additional) => {
    statusesList.forEach((status) => {
        if (additional?.[status]) {
            incrementStatistic(statistic, status, additional[status]);
        }
    });
};
