// @allurereport/plugin-agent/dist/harness.js
import { readFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { ENRICHMENT_ACTIONS_BY_CHECK_NAME } from "./guidance.js";
export const DEFAULT_ANTI_DUMMY_CONFIDENCE = 0.75;
const FALLBACK_ACTION = {
    checkName: "*",
    category: "review-manually",
    title: "Review the finding directly",
    guidance: "Inspect the linked markdown and follow the remediation hint before rerunning.",
};
export const AGENT_ENRICHMENT_ACTIONS = Object.fromEntries(Object.entries(ENRICHMENT_ACTIONS_BY_CHECK_NAME).map(([checkName, action]) => [checkName, { checkName, ...action }]));
export const SCOPE_REJECTING_CHECKS = [
    "missing-expected-test",
    "missing-expected-prefix",
    "missing-expected-environment",
    "unexpected-environment",
    "forbidden-selector-match",
    "unexpected-test",
];
export const ITERATION_REQUIRED_CHECKS = [
    "invalid-expectations-file",
    "no-visible-tests",
    "runner-failures-outside-logical-results",
    "missing-expected-label-selector",
    "metadata-mismatch",
    "history-id-collision",
    "failed-without-useful-steps",
    "failed-without-attachments",
    "nontrivial-run-with-empty-trace",
    "retries-without-new-evidence",
    "passed-without-observable-evidence",
];
export const ANTI_DUMMY_CHECKS = ["noop-dominated-steps"];
const SEVERITY_ORDER = {
    high: 0,
    warning: 1,
    info: 2,
};
const IMPACT_ORDER = {
    reject: 0,
    iterate: 1,
    advisory: 2,
};
const uniqueValues = (values) => Array.from(new Set(values));
const normalizeStringArray = (value) => {
    if (typeof value === "string") {
        return value.length ? [value] : [];
    }
    if (!Array.isArray(value)) {
        return [];
    }
    return uniqueValues(value.filter((item) => typeof item === "string" && item.length > 0));
};
const normalizeLabelValues = (value) => Object.fromEntries(Object.entries(value ?? {}).flatMap(([name, rawValue]) => {
    const values = normalizeStringArray(rawValue);
    return values.length ? [[name, values]] : [];
}));
const toExpectationSelector = (input) => {
    if (!input) {
        return undefined;
    }
    const environments = normalizeStringArray(input.environments);
    const fullNames = normalizeStringArray(input.fullNames);
    const fullNamePrefixes = normalizeStringArray(input.fullNamePrefixes);
    const labelValues = normalizeLabelValues(input.labelValues);
    if (!environments.length && !fullNames.length && !fullNamePrefixes.length && !Object.keys(labelValues).length) {
        return undefined;
    }
    return {
        ...(environments.length ? { environments } : {}),
        ...(fullNames.length ? { full_names: fullNames } : {}),
        ...(fullNamePrefixes.length ? { full_name_prefixes: fullNamePrefixes } : {}),
        ...(Object.keys(labelValues).length ? { label_values: labelValues } : {}),
    };
};
const readJson = async (path) => JSON.parse(await readFile(path, "utf-8"));
const readJsonl = async (path) => (await readFile(path, "utf-8"))
    .trim()
    .split("\n")
    .filter(Boolean)
    .map((line) => JSON.parse(line));
const countByActionCategory = (items) => items.reduce((counts, item) => {
    counts[item.category] += 1;
    return counts;
}, {
    "bootstrap-allure": 0,
    "narrow-test-scope": 0,
    "repair-test-metadata": 0,
    "add-meaningful-steps": 0,
    "add-test-attachments": 0,
    "add-retry-diagnostics": 0,
    "collapse-low-signal-trace": 0,
    "review-manually": 0,
});
const sortPlan = (items) => [...items].sort((left, right) => {
    const byImpact = IMPACT_ORDER[left.acceptanceImpact] - IMPACT_ORDER[right.acceptanceImpact];
    if (byImpact !== 0) {
        return byImpact;
    }
    const bySeverity = SEVERITY_ORDER[left.severity] - SEVERITY_ORDER[right.severity];
    if (bySeverity !== 0) {
        return bySeverity;
    }
    const byCategory = left.category.localeCompare(right.category);
    if (byCategory !== 0) {
        return byCategory;
    }
    const bySubject = left.subject.localeCompare(right.subject);
    if (bySubject !== 0) {
        return bySubject;
    }
    return left.checkName.localeCompare(right.checkName);
});
const impactForFinding = (finding, antiDummyConfidenceThreshold) => {
    if (SCOPE_REJECTING_CHECKS.includes(finding.check_name)) {
        return "reject";
    }
    if (ANTI_DUMMY_CHECKS.includes(finding.check_name) &&
        (finding.confidence ?? 0) >= antiDummyConfidenceThreshold) {
        return "reject";
    }
    if (ITERATION_REQUIRED_CHECKS.includes(finding.check_name)) {
        return "iterate";
    }
    if (finding.severity === "high") {
        return "iterate";
    }
    return "advisory";
};
export const buildAgentExpectations = (input) => {
    const expected = toExpectationSelector(input.target ?? input.expected);
    const forbidden = toExpectationSelector(input.forbidden);
    const notes = normalizeStringArray(input.notes);
    return {
        ...(input.goal ? { goal: input.goal } : {}),
        ...(input.taskId ? { task_id: input.taskId } : {}),
        ...(expected ? { expected } : {}),
        ...(forbidden ? { forbidden } : {}),
        ...(notes.length ? { notes } : {}),
    };
};
export const mapFindingToEnrichmentAction = (finding) => {
    const checkName = typeof finding === "string" ? finding : finding.check_name;
    const mapped = AGENT_ENRICHMENT_ACTIONS[checkName];
    return mapped ?? { ...FALLBACK_ACTION, checkName };
};
export const loadAgentOutput = async (outputDir) => {
    const absoluteOutputDir = resolve(outputDir);
    const run = await readJson(join(absoluteOutputDir, "manifest", "run.json"));
    const tests = await readJsonl(join(absoluteOutputDir, "manifest", "tests.jsonl"));
    const findings = await readJsonl(join(absoluteOutputDir, "manifest", "findings.jsonl"));
    const expected = run.paths.expected_manifest && run.expectations_present
        ? await readJson(join(absoluteOutputDir, run.paths.expected_manifest))
        : undefined;
    return {
        outputDir: absoluteOutputDir,
        run,
        tests,
        findings,
        expected,
    };
};
export const planAgentEnrichmentReview = (output, options = {}) => {
    const antiDummyConfidenceThreshold = options.antiDummyConfidenceThreshold ?? DEFAULT_ANTI_DUMMY_CONFIDENCE;
    const requireExpectationsForAcceptance = options.requireExpectationsForAcceptance ?? true;
    const testsByPath = new Map(output.tests.map((test) => [test.markdown_path, test]));
    const plan = sortPlan(output.findings.map((finding) => {
        const action = mapFindingToEnrichmentAction(finding);
        const matchedTest = testsByPath.get(finding.subject);
        return {
            ...action,
            subject: finding.subject,
            subjectType: finding.subject === "run" ? "run" : "test",
            severity: finding.severity,
            message: finding.message,
            explanation: finding.explanation,
            remediationHint: finding.remediation_hint,
            evidencePaths: finding.evidence_paths,
            expectedReference: finding.expected_reference,
            confidence: finding.confidence,
            acceptanceImpact: impactForFinding(finding, antiDummyConfidenceThreshold),
            fullName: matchedTest?.full_name,
            markdownPath: matchedTest?.markdown_path,
            scopeMatch: matchedTest?.scope_match,
        };
    }));
    const rejecting = plan.filter((item) => item.acceptanceImpact === "reject");
    const iterate = plan.filter((item) => item.acceptanceImpact === "iterate");
    const advisory = plan.filter((item) => item.acceptanceImpact === "advisory");
    const notes = [];
    if (!output.run.expectations_present) {
        notes.push("Generate ALLURE_AGENT_EXPECTATIONS before the next enrichment iteration so scope checks are comparable.");
    }
    if (rejecting.some((item) => item.checkName === "noop-dominated-steps")) {
        notes.push("Reject noop-dominated enrichment: keep only steps tied to real actions or checks, and use real runtime attachments instead of placeholders.");
    }
    if (rejecting.some((item) => SCOPE_REJECTING_CHECKS.includes(item.checkName))) {
        notes.push("Reject scope drift: rerun only the intended tests and keep forbidden scope in the expectations file.");
    }
    if (!notes.length && !iterate.length && output.run.expectations_present) {
        notes.push("Scope matched expectations and no blocking evidence gaps or anti-dummy findings remain.");
    }
    const status = rejecting.length
        ? "reject"
        : requireExpectationsForAcceptance && !output.run.expectations_present
            ? "iterate"
            : iterate.length
                ? "iterate"
                : "accept";
    return {
        status,
        outputDir: output.outputDir,
        expectationsPresent: output.run.expectations_present,
        expected: output.expected,
        summary: {
            totalTests: output.tests.length,
            totalFindings: output.findings.length,
            planItems: plan.length,
            rejectingItems: rejecting.length,
            iterationItems: iterate.length,
            advisoryItems: advisory.length,
            countsByActionCategory: countByActionCategory(plan),
        },
        notes,
        plan,
        rejecting,
        iterate,
        advisory,
        rerun: {
            requiresExpectations: requireExpectationsForAcceptance,
            useExistingExpectations: output.run.expectations_present,
            targetedTests: uniqueValues(plan.flatMap((item) => (item.fullName ? [item.fullName] : []))),
        },
    };
};
export const reviewAgentOutput = async (outputDir, options) => planAgentEnrichmentReview(await loadAgentOutput(outputDir), options);
