// @allurereport/core-api/dist/utils/environment.js
export const DEFAULT_ENVIRONMENT = "default";
export const MAX_ENVIRONMENT_NAME_LENGTH = 64;
export const MAX_ENVIRONMENT_ID_LENGTH = 64;
export const DEFAULT_ENVIRONMENT_IDENTITY = {
    id: DEFAULT_ENVIRONMENT,
    name: DEFAULT_ENVIRONMENT,
};
const hasControlChars = (value) => {
    for (let i = 0; i < value.length; i++) {
        const code = value.charCodeAt(i);
        if (code <= 0x1f || (code >= 0x7f && code <= 0x9f)) {
            return true;
        }
    }
    return false;
};
const hasPathLikeSegments = (value) => {
    if (value.includes("/") || value.includes("\\")) {
        return true;
    }
    return value === "." || value === "..";
};
export const validateEnvironmentName = (name) => {
    if (typeof name !== "string") {
        return { valid: false, reason: "name must be a string" };
    }
    const normalized = name.trim();
    if (normalized.length === 0) {
        return { valid: false, reason: "name must not be empty" };
    }
    if (normalized.length > MAX_ENVIRONMENT_NAME_LENGTH) {
        return {
            valid: false,
            reason: `name must not exceed ${MAX_ENVIRONMENT_NAME_LENGTH} characters`,
        };
    }
    if (hasControlChars(normalized)) {
        return { valid: false, reason: "name must not contain control characters" };
    }
    if (hasPathLikeSegments(normalized)) {
        return { valid: false, reason: "name must not contain path-like segments" };
    }
    return { valid: true, normalized };
};
export const validateEnvironmentId = (environmentId) => {
    if (typeof environmentId !== "string") {
        return { valid: false, reason: "id must be a string" };
    }
    const normalized = environmentId.trim();
    if (normalized.length === 0) {
        return { valid: false, reason: "id must not be empty" };
    }
    if (normalized.length > MAX_ENVIRONMENT_ID_LENGTH) {
        return {
            valid: false,
            reason: `id must not exceed ${MAX_ENVIRONMENT_ID_LENGTH} characters`,
        };
    }
    if (!/^[A-Za-z0-9_-]+$/.test(normalized)) {
        return {
            valid: false,
            reason: "id must contain only latin letters, digits, underscores, and hyphens",
        };
    }
    return { valid: true, normalized };
};
export const assertValidEnvironmentName = (name, source = "environment name") => {
    const validationResult = validateEnvironmentName(name);
    if (!validationResult.valid) {
        throw new Error(`Invalid ${source} ${JSON.stringify(name)}: ${validationResult.reason}`);
    }
    return validationResult.normalized;
};
export const formatNormalizedEnvironmentCollision = (sourcePath, normalized, originalKeys) => `${sourcePath}: normalized key ${JSON.stringify(normalized)} is produced by original keys [${originalKeys.map((key) => JSON.stringify(key)).join(",")}]`;
export const getRealEnvsCount = (group) => {
    const { testResultsByEnv = {} } = group ?? {};
    const envsCount = Object.keys(testResultsByEnv).length;
    if (envsCount <= 1 && DEFAULT_ENVIRONMENT in testResultsByEnv) {
        return 0;
    }
    return envsCount;
};
