// @allurereport/core-api/dist/utils/history.js
import { createHash } from "node:crypto";
import { fallbackTestCaseIdLabelName } from "../constants.js";
import { findLastByLabelName } from "./label.js";
const md5 = (data) => createHash("md5").update(data).digest("hex");
const parametersCompare = (a, b) => {
    return (a.name ?? "").localeCompare(b.name ?? "") || (a.value ?? "").localeCompare(b.value ?? "");
};
export const stringifyHistoryParams = (parameters = []) => {
    return [...parameters]
        .filter((parameter) => !parameter?.excluded)
        .sort(parametersCompare)
        .map((parameter) => `${parameter.name}:${parameter.value}`)
        .join(",");
};
export const getFallbackHistoryId = (tr) => {
    const fallbackTestCaseId = findLastByLabelName(tr.labels ?? [], fallbackTestCaseIdLabelName);
    if (!fallbackTestCaseId) {
        return undefined;
    }
    return `${fallbackTestCaseId}.${md5(stringifyHistoryParams(tr.parameters ?? []))}`;
};
export const getHistoryIdCandidates = (tr) => {
    const result = [];
    if (tr.historyId) {
        result.push(tr.historyId);
    }
    const fallbackHistoryId = getFallbackHistoryId(tr);
    if (fallbackHistoryId && !result.includes(fallbackHistoryId)) {
        result.push(fallbackHistoryId);
    }
    return result;
};
export const filterUnknownByKnownIssues = (trs, knownIssueHistoryIds) => {
    return trs.filter((tr) => {
        const historyIdCandidates = getHistoryIdCandidates(tr);
        if (historyIdCandidates.length === 0) {
            return true;
        }
        return historyIdCandidates.every((historyId) => !knownIssueHistoryIds.has(historyId));
    });
};
export const normalizeHistoryDataPointUrls = (historyDataPoint) => {
    const { url } = historyDataPoint;
    if (!url) {
        return historyDataPoint;
    }
    let testResults = historyDataPoint.testResults;
    for (const [historyId, historyTestResult] of Object.entries(historyDataPoint.testResults)) {
        if (historyTestResult.url) {
            continue;
        }
        if (testResults === historyDataPoint.testResults) {
            testResults = { ...historyDataPoint.testResults };
        }
        testResults[historyId] = {
            ...historyTestResult,
            url,
        };
    }
    if (testResults === historyDataPoint.testResults) {
        return historyDataPoint;
    }
    return {
        ...historyDataPoint,
        testResults,
    };
};
export const selectHistoryTestResults = (historyDataPoints, historyIdCandidates) => {
    if (historyIdCandidates.length === 0) {
        return [];
    }
    return historyDataPoints.reduce((acc, historyDataPoint) => {
        for (const historyId of historyIdCandidates) {
            const historyTestResult = historyDataPoint.testResults[historyId];
            if (!historyTestResult) {
                continue;
            }
            acc.push(historyTestResult);
            break;
        }
        return acc;
    }, []);
};
export const htrsByTr = (hdps, tr) => {
    if (!tr?.historyId) {
        return [];
    }
    return selectHistoryTestResults(hdps, [tr.historyId]);
};
